{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb498f7b",
   "metadata": {},
   "source": [
    "# Benchmarking RNA-seq DEG Methods with the Dexamethasone Benchmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f3fe067d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import statistics\n",
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "from sklearn import preprocessing\n",
    "import warnings\n",
    "import requests\n",
    "import json\n",
    "import time\n",
    "import scipy.stats as ss\n",
    "from itertools import combinations\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from maayanlab_bioinformatics.dge.characteristic_direction import characteristic_direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d8c2280",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pandas2ri.activate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcdbaf65",
   "metadata": {},
   "source": [
    "# Load in Data\n",
    "\n",
    "Using data from GEO from the study, \"The effect of lithium and dexamethasone on fetal rat metatarsal bones transcriptome\" \n",
    "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186104 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cc26d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "meta_df = pd.read_csv('GSE186104_series_matrix.txt', sep=\"\\t\", index_col=0, dtype=str)\n",
    "expr_df = pd.read_csv('GSE186104_cross_tabulation_of_gene_expression.txt', index_col=0, sep=\"\\t\").sort_index()\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: standard cultivation medium'\n",
    "\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36e3e6db",
   "metadata": {},
   "outputs": [],
   "source": [
    "low_expression_threshold = 0.3\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3603da18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5632354</th>\n",
       "      <td>C1: Untreated control replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632355</th>\n",
       "      <td>C2: Untreated control replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632356</th>\n",
       "      <td>C3: Untreated control replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632357</th>\n",
       "      <td>Dex1: Dexamethason treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632358</th>\n",
       "      <td>Dex2: Dexamethason treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632359</th>\n",
       "      <td>Dex3: Dexamethason treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632360</th>\n",
       "      <td>Li1: Lithium treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632361</th>\n",
       "      <td>Li2: Lithium treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632362</th>\n",
       "      <td>Li3: Lithium treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632363</th>\n",
       "      <td>DL1: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632364</th>\n",
       "      <td>DL2: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632365</th>\n",
       "      <td>DL3: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Sample_title  \\\n",
       "Sample_geo_accession                                                      \n",
       "GSM5632354                            C1: Untreated control replicate 1   \n",
       "GSM5632355                            C2: Untreated control replicate 2   \n",
       "GSM5632356                            C3: Untreated control replicate 3   \n",
       "GSM5632357                Dex1: Dexamethason treated sample replicate 1   \n",
       "GSM5632358                Dex2: Dexamethason treated sample replicate 2   \n",
       "GSM5632359                Dex3: Dexamethason treated sample replicate 3   \n",
       "GSM5632360                      Li1: Lithium treated sample replicate 1   \n",
       "GSM5632361                      Li2: Lithium treated sample replicate 2   \n",
       "GSM5632362                      Li3: Lithium treated sample replicate 3   \n",
       "GSM5632363            DL1: Dexamethason and lithium treated sample r...   \n",
       "GSM5632364            DL2: Dexamethason and lithium treated sample r...   \n",
       "GSM5632365            DL3: Dexamethason and lithium treated sample r...   \n",
       "\n",
       "                                             Sample_characteristics_ch1  \n",
       "Sample_geo_accession                                                     \n",
       "GSM5632354                       treatment: standard cultivation medium  \n",
       "GSM5632355                       treatment: standard cultivation medium  \n",
       "GSM5632356                       treatment: standard cultivation medium  \n",
       "GSM5632357            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632358            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632359            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632360             treatment: standard cultivation medium + lithium  \n",
       "GSM5632361             treatment: standard cultivation medium + lithium  \n",
       "GSM5632362             treatment: standard cultivation medium + lithium  \n",
       "GSM5632363            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632364            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632365            treatment: standard cultivation medium + dexam...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "86d112cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "      <th>GSM5632360</th>\n",
       "      <th>GSM5632361</th>\n",
       "      <th>GSM5632362</th>\n",
       "      <th>GSM5632363</th>\n",
       "      <th>GSM5632364</th>\n",
       "      <th>GSM5632365</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>160</td>\n",
       "      <td>60</td>\n",
       "      <td>196</td>\n",
       "      <td>328</td>\n",
       "      <td>274</td>\n",
       "      <td>287</td>\n",
       "      <td>179</td>\n",
       "      <td>188</td>\n",
       "      <td>343</td>\n",
       "      <td>113</td>\n",
       "      <td>93</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>32</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>62</td>\n",
       "      <td>65</td>\n",
       "      <td>36</td>\n",
       "      <td>28</td>\n",
       "      <td>34</td>\n",
       "      <td>68</td>\n",
       "      <td>23</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ml1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>80</td>\n",
       "      <td>36</td>\n",
       "      <td>75</td>\n",
       "      <td>68</td>\n",
       "      <td>70</td>\n",
       "      <td>47</td>\n",
       "      <td>63</td>\n",
       "      <td>44</td>\n",
       "      <td>104</td>\n",
       "      <td>55</td>\n",
       "      <td>69</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>392</td>\n",
       "      <td>199</td>\n",
       "      <td>385</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "      <td>142</td>\n",
       "      <td>189</td>\n",
       "      <td>245</td>\n",
       "      <td>303</td>\n",
       "      <td>269</td>\n",
       "      <td>361</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>374</td>\n",
       "      <td>218</td>\n",
       "      <td>233</td>\n",
       "      <td>144</td>\n",
       "      <td>192</td>\n",
       "      <td>133</td>\n",
       "      <td>156</td>\n",
       "      <td>176</td>\n",
       "      <td>244</td>\n",
       "      <td>210</td>\n",
       "      <td>327</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>611</td>\n",
       "      <td>344</td>\n",
       "      <td>560</td>\n",
       "      <td>222</td>\n",
       "      <td>316</td>\n",
       "      <td>222</td>\n",
       "      <td>286</td>\n",
       "      <td>327</td>\n",
       "      <td>497</td>\n",
       "      <td>369</td>\n",
       "      <td>513</td>\n",
       "      <td>569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>1048</td>\n",
       "      <td>598</td>\n",
       "      <td>922</td>\n",
       "      <td>505</td>\n",
       "      <td>596</td>\n",
       "      <td>399</td>\n",
       "      <td>493</td>\n",
       "      <td>646</td>\n",
       "      <td>954</td>\n",
       "      <td>684</td>\n",
       "      <td>1049</td>\n",
       "      <td>1192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>987</td>\n",
       "      <td>453</td>\n",
       "      <td>863</td>\n",
       "      <td>453</td>\n",
       "      <td>705</td>\n",
       "      <td>387</td>\n",
       "      <td>450</td>\n",
       "      <td>609</td>\n",
       "      <td>798</td>\n",
       "      <td>691</td>\n",
       "      <td>1016</td>\n",
       "      <td>1014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21199 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                         160          60         196         328   \n",
       "A2m                           32          15          19          62   \n",
       "A2ml1                          1           0           0           0   \n",
       "A3galt2                       80          36          75          68   \n",
       "A4galt                         0           0           2           0   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        392         199         385         174   \n",
       "mrpl11                       374         218         233         144   \n",
       "mrpl24                       611         344         560         222   \n",
       "mrpl9                       1048         598         922         505   \n",
       "rnf141                       987         453         863         453   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  GSM5632360  GSM5632361  \\\n",
       "gene                                                                   \n",
       "A1i3                         274         287         179         188   \n",
       "A2m                           65          36          28          34   \n",
       "A2ml1                          2           0           3           0   \n",
       "A3galt2                       70          47          63          44   \n",
       "A4galt                         1           2           0           0   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        185         142         189         245   \n",
       "mrpl11                       192         133         156         176   \n",
       "mrpl24                       316         222         286         327   \n",
       "mrpl9                        596         399         493         646   \n",
       "rnf141                       705         387         450         609   \n",
       "\n",
       "Sample_geo_accession  GSM5632362  GSM5632363  GSM5632364  GSM5632365  \n",
       "gene                                                                  \n",
       "A1i3                         343         113          93         119  \n",
       "A2m                           68          23          29          29  \n",
       "A2ml1                          0           2           4           1  \n",
       "A3galt2                      104          55          69          76  \n",
       "A4galt                         0           0           1           0  \n",
       "...                          ...         ...         ...         ...  \n",
       "l7Rn6                        303         269         361         407  \n",
       "mrpl11                       244         210         327         287  \n",
       "mrpl24                       497         369         513         569  \n",
       "mrpl9                        954         684        1049        1192  \n",
       "rnf141                       798         691        1016        1014  \n",
       "\n",
       "[21199 rows x 12 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff8e58ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def logCPM(data):\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        data = (data/data.sum())*10**6\n",
    "        data = data.fillna(0)\n",
    "        data = np.log2(data+1)\n",
    "\n",
    "    # Return\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b5a843c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize(dataset, current_dataset, logCPM_normalization, log_normalization, z_normalization, q_normalization):\n",
    "    normalization = current_dataset\n",
    "    if logCPM_normalization == True:  \n",
    "        data = dataset[normalization]\n",
    "        normalization += '+logCPM'\n",
    "        dataset[normalization] = logCPM(data)\n",
    "        \n",
    "    if log_normalization == True:    \n",
    "        data = dataset[normalization]\n",
    "        normalization += '+log'\n",
    "        dataset[normalization] = log(data)\n",
    "        \n",
    "    if z_normalization == True:\n",
    "        data = dataset[normalization]\n",
    "        normalization += '+z_norm'    \n",
    "        dataset[normalization] = data.T.apply(ss.zscore, axis=0).T.dropna()\n",
    "\n",
    "    if q_normalization == True:\n",
    "        data = dataset[normalization]\n",
    "        normalization += '+q_norm'\n",
    "        dataset[normalization] = qnormalization(data)\n",
    "    return dataset, normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4e5218a4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dataset, normalization = normalize(dataset, current_dataset, True, False, True, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bf46709c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "      <th>GSM5632360</th>\n",
       "      <th>GSM5632361</th>\n",
       "      <th>GSM5632362</th>\n",
       "      <th>GSM5632363</th>\n",
       "      <th>GSM5632364</th>\n",
       "      <th>GSM5632365</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>-0.823915</td>\n",
       "      <td>-1.203701</td>\n",
       "      <td>-0.394838</td>\n",
       "      <td>1.472828</td>\n",
       "      <td>0.950546</td>\n",
       "      <td>1.391266</td>\n",
       "      <td>0.682852</td>\n",
       "      <td>0.464780</td>\n",
       "      <td>0.636785</td>\n",
       "      <td>-0.661800</td>\n",
       "      <td>-1.357623</td>\n",
       "      <td>-1.157182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>-0.850702</td>\n",
       "      <td>-0.913618</td>\n",
       "      <td>-1.407946</td>\n",
       "      <td>1.707599</td>\n",
       "      <td>1.497647</td>\n",
       "      <td>0.792520</td>\n",
       "      <td>0.368323</td>\n",
       "      <td>0.377260</td>\n",
       "      <td>0.752250</td>\n",
       "      <td>-0.662710</td>\n",
       "      <td>-0.757919</td>\n",
       "      <td>-0.902703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ml1</th>\n",
       "      <td>-0.294967</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>0.920040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>2.292632</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>0.670972</td>\n",
       "      <td>1.308982</td>\n",
       "      <td>-0.265415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>-0.875900</td>\n",
       "      <td>-1.189250</td>\n",
       "      <td>-0.782604</td>\n",
       "      <td>1.542712</td>\n",
       "      <td>1.049973</td>\n",
       "      <td>0.388160</td>\n",
       "      <td>1.659979</td>\n",
       "      <td>-0.508135</td>\n",
       "      <td>0.933154</td>\n",
       "      <td>-0.559917</td>\n",
       "      <td>-0.820001</td>\n",
       "      <td>-0.838169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>1.011257</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>0.792626</td>\n",
       "      <td>2.716666</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>0.285531</td>\n",
       "      <td>-0.600760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>-0.314866</td>\n",
       "      <td>0.016371</td>\n",
       "      <td>0.365216</td>\n",
       "      <td>-0.613465</td>\n",
       "      <td>-1.450318</td>\n",
       "      <td>-1.753019</td>\n",
       "      <td>1.168210</td>\n",
       "      <td>1.839965</td>\n",
       "      <td>-0.805107</td>\n",
       "      <td>0.471590</td>\n",
       "      <td>0.450185</td>\n",
       "      <td>0.625238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>0.682765</td>\n",
       "      <td>1.931779</td>\n",
       "      <td>-2.038182</td>\n",
       "      <td>-0.558921</td>\n",
       "      <td>0.423214</td>\n",
       "      <td>-0.539152</td>\n",
       "      <td>0.762820</td>\n",
       "      <td>0.270977</td>\n",
       "      <td>-0.896703</td>\n",
       "      <td>-0.159605</td>\n",
       "      <td>0.889206</td>\n",
       "      <td>-0.768198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>0.115012</td>\n",
       "      <td>1.517950</td>\n",
       "      <td>0.140863</td>\n",
       "      <td>-2.218022</td>\n",
       "      <td>-0.175757</td>\n",
       "      <td>-1.392579</td>\n",
       "      <td>1.395951</td>\n",
       "      <td>0.835633</td>\n",
       "      <td>0.106610</td>\n",
       "      <td>-0.338141</td>\n",
       "      <td>-0.004243</td>\n",
       "      <td>0.016723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>-0.874443</td>\n",
       "      <td>0.767789</td>\n",
       "      <td>-1.288662</td>\n",
       "      <td>-0.328218</td>\n",
       "      <td>-0.159364</td>\n",
       "      <td>-1.983152</td>\n",
       "      <td>0.546399</td>\n",
       "      <td>1.424708</td>\n",
       "      <td>0.333021</td>\n",
       "      <td>-0.520593</td>\n",
       "      <td>0.898979</td>\n",
       "      <td>1.183536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>-0.587765</td>\n",
       "      <td>-1.164860</td>\n",
       "      <td>-0.978082</td>\n",
       "      <td>-0.571429</td>\n",
       "      <td>1.998474</td>\n",
       "      <td>-1.230443</td>\n",
       "      <td>0.291034</td>\n",
       "      <td>1.288092</td>\n",
       "      <td>-0.648133</td>\n",
       "      <td>0.315527</td>\n",
       "      <td>1.098828</td>\n",
       "      <td>0.188757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21199 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                   -0.823915   -1.203701   -0.394838    1.472828   \n",
       "A2m                    -0.850702   -0.913618   -1.407946    1.707599   \n",
       "A2ml1                  -0.294967   -0.772040   -0.772040   -0.772040   \n",
       "A3galt2                -0.875900   -1.189250   -0.782604    1.542712   \n",
       "A4galt                 -0.600760   -0.600760    1.011257   -0.600760   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                  -0.314866    0.016371    0.365216   -0.613465   \n",
       "mrpl11                  0.682765    1.931779   -2.038182   -0.558921   \n",
       "mrpl24                  0.115012    1.517950    0.140863   -2.218022   \n",
       "mrpl9                  -0.874443    0.767789   -1.288662   -0.328218   \n",
       "rnf141                 -0.587765   -1.164860   -0.978082   -0.571429   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  GSM5632360  GSM5632361  \\\n",
       "gene                                                                   \n",
       "A1i3                    0.950546    1.391266    0.682852    0.464780   \n",
       "A2m                     1.497647    0.792520    0.368323    0.377260   \n",
       "A2ml1                   0.920040   -0.772040    2.292632   -0.772040   \n",
       "A3galt2                 1.049973    0.388160    1.659979   -0.508135   \n",
       "A4galt                  0.792626    2.716666   -0.600760   -0.600760   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                  -1.450318   -1.753019    1.168210    1.839965   \n",
       "mrpl11                  0.423214   -0.539152    0.762820    0.270977   \n",
       "mrpl24                 -0.175757   -1.392579    1.395951    0.835633   \n",
       "mrpl9                  -0.159364   -1.983152    0.546399    1.424708   \n",
       "rnf141                  1.998474   -1.230443    0.291034    1.288092   \n",
       "\n",
       "Sample_geo_accession  GSM5632362  GSM5632363  GSM5632364  GSM5632365  \n",
       "gene                                                                  \n",
       "A1i3                    0.636785   -0.661800   -1.357623   -1.157182  \n",
       "A2m                     0.752250   -0.662710   -0.757919   -0.902703  \n",
       "A2ml1                  -0.772040    0.670972    1.308982   -0.265415  \n",
       "A3galt2                 0.933154   -0.559917   -0.820001   -0.838169  \n",
       "A4galt                 -0.600760   -0.600760    0.285531   -0.600760  \n",
       "...                          ...         ...         ...         ...  \n",
       "l7Rn6                  -0.805107    0.471590    0.450185    0.625238  \n",
       "mrpl11                 -0.896703   -0.159605    0.889206   -0.768198  \n",
       "mrpl24                  0.106610   -0.338141   -0.004243    0.016723  \n",
       "mrpl9                   0.333021   -0.520593    0.898979    1.183536  \n",
       "rnf141                 -0.648133    0.315527    1.098828    0.188757  \n",
       "\n",
       "[21199 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[normalization]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc27b4b4",
   "metadata": {},
   "source": [
    "# Differential Gene Expression\n",
    "Using code adapted from Bulk RNA-seq Analysis pipeline appyter: https://appyters.maayanlab.cloud/Bulk_RNA_seq/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "92c3b12a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'robjects.r(\\'\\'\\'limma <- function(rawcount_dataframe, design_dataframe, filter_genes=FALSE, adjust=\"BH\") {\\n    # Load packages\\n    suppressMessages(require(limma))\\n    suppressMessages(require(edgeR))\\n    # Convert design matrix\\n    design <- as.matrix(design_dataframe)\\n    \\n    # Create DGEList object\\n    dge <- DGEList(counts=rawcount_dataframe)\\n    # Filter genes\\n    if (filter_genes) {\\n        keep <- filterByExpr(dge, design)\\n        dge <- dge[keep,]\\n    }\\n    # Calculate normalization factors\\n    dge <- calcNormFactors(dge)\\n    # Run VOOM\\n    v <- voom(dge, plot=FALSE)\\n    # Fit linear model\\n    fit <- lmFit(v, design)\\n    # Make contrast matrix\\n    cont.matrix <- makeContrasts(de=B-A, levels=design)\\n    # Fit\\n    fit2 <- contrasts.fit(fit, cont.matrix)\\n    # Run DE\\n    fit2 <- eBayes(fit2)\\n    # Get results\\n    limma_dataframe <- topTable(fit2, adjust=adjust, number=nrow(rawcount_dataframe))\\n    \\n    # Return\\n    results <- list(\"limma_dataframe\"= limma_dataframe, \"rownames\"=rownames(limma_dataframe))\\n    return (results)\\n}\\n\\'\\'\\')'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Limma not working\n",
    "\"\"\"robjects.r('''limma <- function(rawcount_dataframe, design_dataframe, filter_genes=FALSE, adjust=\"BH\") {\n",
    "    # Load packages\n",
    "    suppressMessages(require(limma))\n",
    "    suppressMessages(require(edgeR))\n",
    "    # Convert design matrix\n",
    "    design <- as.matrix(design_dataframe)\n",
    "    \n",
    "    # Create DGEList object\n",
    "    dge <- DGEList(counts=rawcount_dataframe)\n",
    "    # Filter genes\n",
    "    if (filter_genes) {\n",
    "        keep <- filterByExpr(dge, design)\n",
    "        dge <- dge[keep,]\n",
    "    }\n",
    "    # Calculate normalization factors\n",
    "    dge <- calcNormFactors(dge)\n",
    "    # Run VOOM\n",
    "    v <- voom(dge, plot=FALSE)\n",
    "    # Fit linear model\n",
    "    fit <- lmFit(v, design)\n",
    "    # Make contrast matrix\n",
    "    cont.matrix <- makeContrasts(de=B-A, levels=design)\n",
    "    # Fit\n",
    "    fit2 <- contrasts.fit(fit, cont.matrix)\n",
    "    # Run DE\n",
    "    fit2 <- eBayes(fit2)\n",
    "    # Get results\n",
    "    limma_dataframe <- topTable(fit2, adjust=adjust, number=nrow(rawcount_dataframe))\n",
    "    \n",
    "    # Return\n",
    "    results <- list(\"limma_dataframe\"= limma_dataframe, \"rownames\"=rownames(limma_dataframe))\n",
    "    return (results)\n",
    "}\n",
    "''')\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d9bcc6d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from the appyter source code\n",
    "def get_signatures(control, treatment, dataset, normalization, method, meta_class_column_name, filter_genes):\n",
    "    tmp_normalization = normalization.replace(\"+z_norm+q_norm\",\"\").replace(\"+z_norm\",\"\")\n",
    "    raw_expr_df = dataset['rawdata']\n",
    "    expr_df = dataset[current_dataset]\n",
    "    if filter_genes == True:\n",
    "        expr_df = dataset['rawdata+filter_genes']\n",
    "        \n",
    "    signatures = dict()\n",
    "\n",
    "    print(control, treatment)\n",
    "    cls1_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==control, :].index.tolist() #control\n",
    "    cls2_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==treatment,:].index.tolist() #case\n",
    "        \n",
    "    signature_label = \" vs. \".join([control, treatment])\n",
    "       \n",
    "    # Limma not working\n",
    "    if method == \"limma\":\n",
    "        limma = robjects.r['limma']\n",
    "\n",
    "        design_dataframe = pd.DataFrame([{'index': x, 'A': int(x in cls1_sample_ids), 'B': int(x in cls2_sample_ids)} for x in raw_expr_df.columns]).set_index('index')\n",
    "\n",
    "        processed_data = {\"expression\": raw_expr_df, 'design': design_dataframe}\n",
    "            \n",
    "        limma_results = pandas2ri.conversion.rpy2py(limma(pandas2ri.conversion.py2rpy(processed_data['expression']), pandas2ri.conversion.py2rpy(processed_data['design']), filter_genes=filter_genes))\n",
    "                        \n",
    "        signature = pd.DataFrame(limma_results[0])\n",
    "        signature.index = limma_results[1]\n",
    "        signature = signature.sort_values(\"t\", ascending=False)\n",
    "            \n",
    "    elif method == \"logFC\":\n",
    "        values = []\n",
    "        for i in range(len(expr_df)):\n",
    "            case_mean = statistics.mean(expr_df.iloc[i][cls2_sample_ids])\n",
    "            control_mean = statistics.mean(expr_df.iloc[i][cls1_sample_ids])\n",
    "            if case_mean == 0 or control_mean == 0:\n",
    "                values.append('NA')\n",
    "            else:\n",
    "                values.append(math.log(case_mean/control_mean,2))\n",
    "\n",
    "        signature = pd.DataFrame(values, columns = ['logFC'])\n",
    "        signature.index = expr_df.index\n",
    "        signature = signature[signature['logFC'] != 'NA']\n",
    "        signature = df.sort_values(\"logFC\", ascending=False)\n",
    "            \n",
    "    elif method == \"characteristic_direction\":\n",
    "        signature = characteristic_direction(dataset[tmp_normalization].loc[:, cls1_sample_ids], dataset[normalization].loc[:, cls2_sample_ids], calculate_sig=True)\n",
    "        signature = signature.sort_values(\"CD-coefficient\", ascending=False)\n",
    "            \n",
    "    signatures[signature_label] = signature\n",
    "\n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2036a595",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "treatment: standard cultivation medium treatment: standard cultivation medium + dexamethason\n"
     ]
    }
   ],
   "source": [
    "# Set method in this variable\n",
    "diff_gex_method = 'characteristic_direction'\n",
    "control = 'treatment: standard cultivation medium'\n",
    "treatment = 'treatment: standard cultivation medium + dexamethason'\n",
    "\n",
    "signatures = get_signatures(control, treatment, dataset, normalization, diff_gex_method, meta_class_column_name, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e8457e07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rbp7</th>\n",
       "      <td>0.016617</td>\n",
       "      <td>-0.088503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LOC102556144</th>\n",
       "      <td>0.016617</td>\n",
       "      <td>-0.139658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selp</th>\n",
       "      <td>0.016136</td>\n",
       "      <td>-0.073505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LOC102550122</th>\n",
       "      <td>0.016067</td>\n",
       "      <td>-0.135190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Slc18a3</th>\n",
       "      <td>0.016039</td>\n",
       "      <td>-0.066825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col1a1</th>\n",
       "      <td>-0.024673</td>\n",
       "      <td>-0.067146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col11a2</th>\n",
       "      <td>-0.025156</td>\n",
       "      <td>-0.066973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col9a2</th>\n",
       "      <td>-0.025866</td>\n",
       "      <td>-0.067782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serpinh1</th>\n",
       "      <td>-0.026813</td>\n",
       "      <td>-0.071877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col2a1</th>\n",
       "      <td>-0.030783</td>\n",
       "      <td>-0.067358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21199 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              CD-coefficient  Significance\n",
       "gene                                      \n",
       "Rbp7                0.016617     -0.088503\n",
       "LOC102556144        0.016617     -0.139658\n",
       "Selp                0.016136     -0.073505\n",
       "LOC102550122        0.016067     -0.135190\n",
       "Slc18a3             0.016039     -0.066825\n",
       "...                      ...           ...\n",
       "Col1a1             -0.024673     -0.067146\n",
       "Col11a2            -0.025156     -0.066973\n",
       "Col9a2             -0.025866     -0.067782\n",
       "Serpinh1           -0.026813     -0.071877\n",
       "Col2a1             -0.030783     -0.067358\n",
       "\n",
       "[21199 rows x 2 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd_df = signatures['treatment: standard cultivation medium vs. treatment: standard cultivation medium + dexamethason']\n",
    "cd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a66f1c23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "treatment: standard cultivation medium treatment: standard cultivation medium + dexamethason\n"
     ]
    }
   ],
   "source": [
    "diff_gex_method = 'logFC'\n",
    "signatures = get_signatures(control, treatment, dataset, normalization, diff_gex_method, meta_class_column_name, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2a7fd936",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cd163</th>\n",
       "      <td>8.885696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orm1</th>\n",
       "      <td>6.087463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calb2</th>\n",
       "      <td>5.260528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cd101</th>\n",
       "      <td>4.813069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LOC100909954</th>\n",
       "      <td>4.778442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pck1</th>\n",
       "      <td>-5.247928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asb2</th>\n",
       "      <td>-5.426265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LOC102550863</th>\n",
       "      <td>-5.554589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RGD1566307</th>\n",
       "      <td>-5.61471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Itgax</th>\n",
       "      <td>-5.70044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19979 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 logFC\n",
       "gene                  \n",
       "Cd163         8.885696\n",
       "Orm1          6.087463\n",
       "Calb2         5.260528\n",
       "Cd101         4.813069\n",
       "LOC100909954  4.778442\n",
       "...                ...\n",
       "Pck1         -5.247928\n",
       "Asb2         -5.426265\n",
       "LOC102550863 -5.554589\n",
       "RGD1566307    -5.61471\n",
       "Itgax         -5.70044\n",
       "\n",
       "[19979 rows x 1 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc_df = signatures['treatment: standard cultivation medium vs. treatment: standard cultivation medium + dexamethason']\n",
    "fc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4a2f7f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd_up_200_genes = list(cd_df.head(200).index)\n",
    "cd_down_200_genes = list(cd_df.tail(200).index)\n",
    "cd_both_400_genes = cd_up_200_genes + cd_down_200_genes\n",
    "logFC_up_200_genes = list(fc_df.head(200).index)\n",
    "logFC_down_200_genes = list(fc_df.tail(200).index)\n",
    "logFC_both_400_genes = logFC_up_200_genes + logFC_down_200_genes\n",
    "\n",
    "gene_lists = [cd_up_200_genes, cd_down_200_genes, cd_both_400_genes, logFC_up_200_genes, logFC_down_200_genes, logFC_both_400_genes]\n",
    "list_names = [\"CD Method: 200 Upregulated Genes\", \"CD Method: 200 Downregulated Genes\", \"CD Method: Combined 400 Genes\",\n",
    "             \"LogFC Method: 200 Upregulated Genes\", \"LogFC Method: 200 Downregulated Genes\", \"LogFC Method: Combined 400 Genes\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "269c9eee",
   "metadata": {},
   "source": [
    "# Enrichment Analysis with Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3abb034c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get Enrichr Results\n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "\n",
    "    all_ranks = []\n",
    "    all_terms = []\n",
    "    all_pvalues =[] \n",
    "    all_adjusted_pvalues = []\n",
    "    library_success = []\n",
    "    short_id = ''\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = 'Example gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise Exception('Error analyzing gene list')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        time.sleep(0.5)\n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        short_id = data[\"shortId\"]\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "        try:\n",
    "            data = json.loads(response.text)\n",
    "            results_df  = pd.DataFrame(data[library_name])\n",
    "            all_ranks.append(list(results_df[0]))\n",
    "            all_terms.append(list(results_df[1]))\n",
    "            all_pvalues.append(list(results_df[2]))\n",
    "            all_adjusted_pvalues.append(list(results_df[6]))\n",
    "            library_success.append(library_name)\n",
    "        except:\n",
    "            print('Error for ' + library_name + ' library')\n",
    "\n",
    "    return([all_ranks,all_terms,all_pvalues,all_adjusted_pvalues,str(short_id),library_success])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e6165cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for gene_list in gene_lists:\n",
    "    result = Enrichr_API(gene_list, ['ChEA_2016'])\n",
    "    results.append(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "2cf21f46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CD Method: 200 Upregulated Genes', 'CD Method: 200 Upregulated Genes', 'CD Method: 200 Downregulated Genes', 'CD Method: 200 Downregulated Genes', 'CD Method: Combined 400 Genes', 'CD Method: Combined 400 Genes', 'LogFC Method: 200 Upregulated Genes', 'LogFC Method: 200 Upregulated Genes', 'LogFC Method: 200 Downregulated Genes', 'LogFC Method: 200 Downregulated Genes', 'LogFC Method: Combined 400 Genes', 'LogFC Method: Combined 400 Genes'] ['NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human', 'NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human', 'NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human', 'NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human', 'NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human', 'NR3C1 23031785 ChIP-Seq PC12 Mouse', 'NR3C1 21868756 ChIP-Seq MCF10A Human'] [90, 117, 348, 416, 202, 283, 42, 83, 126, 223, 65, 137] [0.955186110169955, 0.9729289929795114, 0.4369210679488471, 0.582927046873835, 0.8233451578185339, 0.9150257082587666, 0.5721541099309007, 0.8043515446028175, 0.9013807866553136, 0.9729289929795114, 0.8233451578185339, 0.9687220897713261]\n"
     ]
    }
   ],
   "source": [
    "# Extract NR3C1 rankings\n",
    "# Initialize lists for storing NR3C1 information\n",
    "gene_set = []\n",
    "names = []\n",
    "ranks = []\n",
    "p_val = []\n",
    "\n",
    "# Iterate over each result\n",
    "for i in range(len(results)):\n",
    "    # Within each gene set, iterate over the transcription factor names\n",
    "    for j in range(len(results[i][1][0])):\n",
    "        # If NR3C1 is found, add the information to the lists\n",
    "        if 'NR3C1' in results[i][1][0][j]:\n",
    "            names.append(results[i][1][0][j])\n",
    "            ranks.append(results[i][0][0][j])\n",
    "            p_val.append(results[i][2][0][j])\n",
    "            gene_set.append(list_names[i])\n",
    "\n",
    "print(gene_set, names, ranks, p_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d94c2217",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Set</th>\n",
       "      <th>Name</th>\n",
       "      <th>Rank</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CD Method: 200 Upregulated Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>90</td>\n",
       "      <td>0.955186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD Method: 200 Upregulated Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>117</td>\n",
       "      <td>0.972929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD Method: 200 Downregulated Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>348</td>\n",
       "      <td>0.436921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD Method: 200 Downregulated Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>416</td>\n",
       "      <td>0.582927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD Method: Combined 400 Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>202</td>\n",
       "      <td>0.823345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CD Method: Combined 400 Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>283</td>\n",
       "      <td>0.915026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LogFC Method: 200 Upregulated Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>42</td>\n",
       "      <td>0.572154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LogFC Method: 200 Upregulated Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>83</td>\n",
       "      <td>0.804352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LogFC Method: 200 Downregulated Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>126</td>\n",
       "      <td>0.901381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LogFC Method: 200 Downregulated Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>223</td>\n",
       "      <td>0.972929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LogFC Method: Combined 400 Genes</td>\n",
       "      <td>NR3C1 23031785 ChIP-Seq PC12 Mouse</td>\n",
       "      <td>65</td>\n",
       "      <td>0.823345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LogFC Method: Combined 400 Genes</td>\n",
       "      <td>NR3C1 21868756 ChIP-Seq MCF10A Human</td>\n",
       "      <td>137</td>\n",
       "      <td>0.968722</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Gene Set  \\\n",
       "0        CD Method: 200 Upregulated Genes   \n",
       "1        CD Method: 200 Upregulated Genes   \n",
       "2      CD Method: 200 Downregulated Genes   \n",
       "3      CD Method: 200 Downregulated Genes   \n",
       "4           CD Method: Combined 400 Genes   \n",
       "5           CD Method: Combined 400 Genes   \n",
       "6     LogFC Method: 200 Upregulated Genes   \n",
       "7     LogFC Method: 200 Upregulated Genes   \n",
       "8   LogFC Method: 200 Downregulated Genes   \n",
       "9   LogFC Method: 200 Downregulated Genes   \n",
       "10       LogFC Method: Combined 400 Genes   \n",
       "11       LogFC Method: Combined 400 Genes   \n",
       "\n",
       "                                    Name  Rank   p-value  \n",
       "0     NR3C1 23031785 ChIP-Seq PC12 Mouse    90  0.955186  \n",
       "1   NR3C1 21868756 ChIP-Seq MCF10A Human   117  0.972929  \n",
       "2     NR3C1 23031785 ChIP-Seq PC12 Mouse   348  0.436921  \n",
       "3   NR3C1 21868756 ChIP-Seq MCF10A Human   416  0.582927  \n",
       "4     NR3C1 23031785 ChIP-Seq PC12 Mouse   202  0.823345  \n",
       "5   NR3C1 21868756 ChIP-Seq MCF10A Human   283  0.915026  \n",
       "6     NR3C1 23031785 ChIP-Seq PC12 Mouse    42  0.572154  \n",
       "7   NR3C1 21868756 ChIP-Seq MCF10A Human    83  0.804352  \n",
       "8     NR3C1 23031785 ChIP-Seq PC12 Mouse   126  0.901381  \n",
       "9   NR3C1 21868756 ChIP-Seq MCF10A Human   223  0.972929  \n",
       "10    NR3C1 23031785 ChIP-Seq PC12 Mouse    65  0.823345  \n",
       "11  NR3C1 21868756 ChIP-Seq MCF10A Human   137  0.968722  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a display dataframe\n",
    "df = pd.DataFrame(list(zip(gene_set, names, ranks, p_val)),\n",
    "                 columns = ['Gene Set', 'Name','Rank','p-value'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07b8cabc",
   "metadata": {},
   "source": [
    "# Comparing methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "49bd1080",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare rankings/methods"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv3.9",
   "language": "python",
   "name": "venv3.9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
