{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de55cbab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import dependencies\n",
    "import pandas as pd\n",
    "import statistics\n",
    "import math\n",
    "from maayanlab_bioinformatics.dge.characteristic_direction import characteristic_direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c50adeac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for computing signatures with characteristic direction\n",
    "def cd_signature(control, treatment, dataset, normalization, meta_class_column_name):\n",
    "    tmp_normalization = normalization\n",
    "    \n",
    "    signatures = dict()\n",
    "    \n",
    "    signature_label = normalization\n",
    "    \n",
    "    cls1_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==control, :].index.tolist() #control\n",
    "    cls2_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==treatment,:].index.tolist() #case\n",
    "    \n",
    "    signature = characteristic_direction(dataset[tmp_normalization].loc[:, cls1_sample_ids], dataset[normalization].loc[:, cls2_sample_ids], calculate_sig=True)\n",
    "    signature = signature.sort_values(\"CD-coefficient\", ascending=False)\n",
    "    \n",
    "    signatures[signature_label] = signature\n",
    "    \n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f248291",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for computing signatures with characteristic direction\n",
    "def logFC(control, treatment, dataset, normalization, meta_class_column_name):\n",
    "    tmp_normalization = normalization\n",
    "    \n",
    "    signatures = dict()\n",
    "    \n",
    "    signature_label = normalization\n",
    "    \n",
    "    cls1_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==control, :].index.tolist() #control\n",
    "    cls2_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==treatment,:].index.tolist() #case\n",
    "    \n",
    "    values = []\n",
    "    for i in range(len(expr_df)):\n",
    "        case_mean = statistics.mean(expr_df.iloc[i][cls2_sample_ids])\n",
    "        control_mean = statistics.mean(expr_df.iloc[i][cls1_sample_ids])\n",
    "        if case_mean == 0 or control_mean == 0:\n",
    "            values.append('NA')\n",
    "        else:\n",
    "            values.append(math.log(case_mean/control_mean,2))\n",
    "\n",
    "    signature = pd.DataFrame(values, columns = ['logFC'])\n",
    "    signature.index = expr_df.index\n",
    "    signature = signature[signature['logFC'] != 'NA']\n",
    "    signature = signature.sort_values(\"logFC\", ascending=False)\n",
    "    \n",
    "    signatures[signature_label] = signature\n",
    "    \n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb998ce0",
   "metadata": {},
   "source": [
    "# GSE159084"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4bfa4a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE159084_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE159084_Gene_Expression_Profiling.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('gene_short_name')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "743b1684",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_title</th>\n",
       "      <th>CON1_FPKM</th>\n",
       "      <th>CON2_FPKM</th>\n",
       "      <th>CON3_FPKM</th>\n",
       "      <th>CON4_FPKM</th>\n",
       "      <th>DEX1_FPKM</th>\n",
       "      <th>DEX2_FPKM</th>\n",
       "      <th>DEX3_FPKM</th>\n",
       "      <th>DEX4_FPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6.088300</td>\n",
       "      <td>4.693950</td>\n",
       "      <td>6.546880</td>\n",
       "      <td>5.157240</td>\n",
       "      <td>4.890360</td>\n",
       "      <td>3.303080</td>\n",
       "      <td>3.485380</td>\n",
       "      <td>0.823810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.613928</td>\n",
       "      <td>0.189342</td>\n",
       "      <td>0.404009</td>\n",
       "      <td>0.097095</td>\n",
       "      <td>0.377022</td>\n",
       "      <td>0.629754</td>\n",
       "      <td>0.287443</td>\n",
       "      <td>0.609918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>995.610000</td>\n",
       "      <td>571.503000</td>\n",
       "      <td>690.374000</td>\n",
       "      <td>637.734000</td>\n",
       "      <td>672.273000</td>\n",
       "      <td>678.245000</td>\n",
       "      <td>649.617000</td>\n",
       "      <td>354.162000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10.451900</td>\n",
       "      <td>16.212900</td>\n",
       "      <td>6.084100</td>\n",
       "      <td>7.135920</td>\n",
       "      <td>10.153800</td>\n",
       "      <td>13.068000</td>\n",
       "      <td>11.227300</td>\n",
       "      <td>10.321400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.369604</td>\n",
       "      <td>0.896261</td>\n",
       "      <td>1.802320</td>\n",
       "      <td>0.072348</td>\n",
       "      <td>1.011430</td>\n",
       "      <td>1.205760</td>\n",
       "      <td>0.319378</td>\n",
       "      <td>0.890901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22551</th>\n",
       "      <td>0.575961</td>\n",
       "      <td>0.372559</td>\n",
       "      <td>0.342286</td>\n",
       "      <td>0.359761</td>\n",
       "      <td>2.561410</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197913</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22552</th>\n",
       "      <td>1.748870</td>\n",
       "      <td>1.560420</td>\n",
       "      <td>1.899630</td>\n",
       "      <td>2.088580</td>\n",
       "      <td>1.539570</td>\n",
       "      <td>1.206660</td>\n",
       "      <td>0.477881</td>\n",
       "      <td>1.144170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22555</th>\n",
       "      <td>1.957540</td>\n",
       "      <td>1.818660</td>\n",
       "      <td>2.289500</td>\n",
       "      <td>1.227500</td>\n",
       "      <td>2.730500</td>\n",
       "      <td>3.532190</td>\n",
       "      <td>2.087320</td>\n",
       "      <td>3.345870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22556</th>\n",
       "      <td>4.931240</td>\n",
       "      <td>4.676480</td>\n",
       "      <td>6.855870</td>\n",
       "      <td>5.807760</td>\n",
       "      <td>4.910870</td>\n",
       "      <td>5.717560</td>\n",
       "      <td>3.574900</td>\n",
       "      <td>1.565490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22562</th>\n",
       "      <td>0.864946</td>\n",
       "      <td>0.609100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.405778</td>\n",
       "      <td>0.546295</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.098627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11610 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_title   CON1_FPKM   CON2_FPKM   CON3_FPKM   CON4_FPKM   DEX1_FPKM  \\\n",
       "0               6.088300    4.693950    6.546880    5.157240    4.890360   \n",
       "2               0.613928    0.189342    0.404009    0.097095    0.377022   \n",
       "4             995.610000  571.503000  690.374000  637.734000  672.273000   \n",
       "5              10.451900   16.212900    6.084100    7.135920   10.153800   \n",
       "6               0.369604    0.896261    1.802320    0.072348    1.011430   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "22551           0.575961    0.372559    0.342286    0.359761    2.561410   \n",
       "22552           1.748870    1.560420    1.899630    2.088580    1.539570   \n",
       "22555           1.957540    1.818660    2.289500    1.227500    2.730500   \n",
       "22556           4.931240    4.676480    6.855870    5.807760    4.910870   \n",
       "22562           0.864946    0.609100    0.000000    0.000000    0.405778   \n",
       "\n",
       "Sample_title   DEX2_FPKM   DEX3_FPKM   DEX4_FPKM  \n",
       "0               3.303080    3.485380    0.823810  \n",
       "2               0.629754    0.287443    0.609918  \n",
       "4             678.245000  649.617000  354.162000  \n",
       "5              13.068000   11.227300   10.321400  \n",
       "6               1.205760    0.319378    0.890901  \n",
       "...                  ...         ...         ...  \n",
       "22551           0.000000    0.197913    0.000000  \n",
       "22552           1.206660    0.477881    1.144170  \n",
       "22555           3.532190    2.087320    3.345870  \n",
       "22556           5.717560    3.574900    1.565490  \n",
       "22562           0.546295    0.000086    0.098627  \n",
       "\n",
       "[11610 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5a1c045a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE159084.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b773ee88",
   "metadata": {},
   "source": [
    "# GSE163065"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "75a32ec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE163065_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE163065_counts_RPMs_RPKMs.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('GeneName')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "75538fe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_title</th>\n",
       "      <th>NA-SD-19</th>\n",
       "      <th>NA-SD-23</th>\n",
       "      <th>NA-SD-27</th>\n",
       "      <th>NA-SD-21</th>\n",
       "      <th>NA-SD-25</th>\n",
       "      <th>NA-SD-29</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>10.249212</td>\n",
       "      <td>8.887646</td>\n",
       "      <td>8.911430</td>\n",
       "      <td>8.453046</td>\n",
       "      <td>9.798075</td>\n",
       "      <td>9.361375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.450035</td>\n",
       "      <td>0.629510</td>\n",
       "      <td>0.600657</td>\n",
       "      <td>0.473945</td>\n",
       "      <td>0.514389</td>\n",
       "      <td>0.626822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>6.792555</td>\n",
       "      <td>6.689296</td>\n",
       "      <td>7.051307</td>\n",
       "      <td>2.790350</td>\n",
       "      <td>2.573155</td>\n",
       "      <td>2.951682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>5.297814</td>\n",
       "      <td>5.220476</td>\n",
       "      <td>5.595217</td>\n",
       "      <td>15.850395</td>\n",
       "      <td>15.601677</td>\n",
       "      <td>16.558084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <td>1.024511</td>\n",
       "      <td>0.997721</td>\n",
       "      <td>1.038536</td>\n",
       "      <td>1.208253</td>\n",
       "      <td>1.456840</td>\n",
       "      <td>1.212021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ts1</th>\n",
       "      <td>8.196085</td>\n",
       "      <td>8.529308</td>\n",
       "      <td>43.756983</td>\n",
       "      <td>5.269952</td>\n",
       "      <td>3.798602</td>\n",
       "      <td>8.462547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tt</th>\n",
       "      <td>68.993909</td>\n",
       "      <td>62.114822</td>\n",
       "      <td>78.370716</td>\n",
       "      <td>84.665324</td>\n",
       "      <td>82.753702</td>\n",
       "      <td>105.425325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tw</th>\n",
       "      <td>225.698161</td>\n",
       "      <td>222.107544</td>\n",
       "      <td>274.787321</td>\n",
       "      <td>226.588286</td>\n",
       "      <td>235.622359</td>\n",
       "      <td>265.109552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ty</th>\n",
       "      <td>0.733978</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.542726</td>\n",
       "      <td>0.601845</td>\n",
       "      <td>0.562268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s189</th>\n",
       "      <td>0.215686</td>\n",
       "      <td>1.290619</td>\n",
       "      <td>0.863625</td>\n",
       "      <td>0.797427</td>\n",
       "      <td>0.884290</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14703 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_title     NA-SD-19    NA-SD-23    NA-SD-27    NA-SD-21    NA-SD-25  \\\n",
       "GeneName                                                                    \n",
       "0610009B22Rik   10.249212    8.887646    8.911430    8.453046    9.798075   \n",
       "0610009E02Rik    0.450035    0.629510    0.600657    0.473945    0.514389   \n",
       "0610009L18Rik    6.792555    6.689296    7.051307    2.790350    2.573155   \n",
       "0610010F05Rik    5.297814    5.220476    5.595217   15.850395   15.601677   \n",
       "0610010K14Rik    1.024511    0.997721    1.038536    1.208253    1.456840   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "mt-Ts1           8.196085    8.529308   43.756983    5.269952    3.798602   \n",
       "mt-Tt           68.993909   62.114822   78.370716   84.665324   82.753702   \n",
       "mt-Tw          225.698161  222.107544  274.787321  226.588286  235.622359   \n",
       "mt-Ty            0.733978    0.000000    0.000000    0.542726    0.601845   \n",
       "n-R5s189         0.215686    1.290619    0.863625    0.797427    0.884290   \n",
       "\n",
       "Sample_title     NA-SD-29  \n",
       "GeneName                   \n",
       "0610009B22Rik    9.361375  \n",
       "0610009E02Rik    0.626822  \n",
       "0610009L18Rik    2.951682  \n",
       "0610010F05Rik   16.558084  \n",
       "0610010K14Rik    1.212021  \n",
       "...                   ...  \n",
       "mt-Ts1           8.462547  \n",
       "mt-Tt          105.425325  \n",
       "mt-Tw          265.109552  \n",
       "mt-Ty            0.562268  \n",
       "n-R5s189         0.000000  \n",
       "\n",
       "[14703 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bf9bf56b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE163065.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157d5c30",
   "metadata": {},
   "source": [
    "# GSE186950"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "06188df7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE186950_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE186950_NormCount.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('Gene_Name')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fcd78d36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_title</th>\n",
       "      <th>AF25</th>\n",
       "      <th>AF26</th>\n",
       "      <th>AF29</th>\n",
       "      <th>AF25DEX</th>\n",
       "      <th>AF26DEX</th>\n",
       "      <th>AF29DEX</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>101.238412</td>\n",
       "      <td>69.133212</td>\n",
       "      <td>85.371425</td>\n",
       "      <td>75.057577</td>\n",
       "      <td>76.011260</td>\n",
       "      <td>57.365282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>947.239407</td>\n",
       "      <td>974.232499</td>\n",
       "      <td>1287.837024</td>\n",
       "      <td>792.393564</td>\n",
       "      <td>698.353451</td>\n",
       "      <td>510.029505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>8.803340</td>\n",
       "      <td>24.560483</td>\n",
       "      <td>5.449240</td>\n",
       "      <td>3.216753</td>\n",
       "      <td>14.252111</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>572.217114</td>\n",
       "      <td>402.973854</td>\n",
       "      <td>455.011530</td>\n",
       "      <td>494.307758</td>\n",
       "      <td>597.400996</td>\n",
       "      <td>493.341423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>492.106718</td>\n",
       "      <td>352.033592</td>\n",
       "      <td>316.055913</td>\n",
       "      <td>551.137066</td>\n",
       "      <td>477.445726</td>\n",
       "      <td>398.427957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDA</th>\n",
       "      <td>47.538037</td>\n",
       "      <td>57.307794</td>\n",
       "      <td>35.420059</td>\n",
       "      <td>28.950780</td>\n",
       "      <td>45.131686</td>\n",
       "      <td>32.333159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>85.392400</td>\n",
       "      <td>140.085719</td>\n",
       "      <td>103.535558</td>\n",
       "      <td>60.046062</td>\n",
       "      <td>67.697528</td>\n",
       "      <td>61.537302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>450.731019</td>\n",
       "      <td>448.456230</td>\n",
       "      <td>479.533109</td>\n",
       "      <td>572.582088</td>\n",
       "      <td>516.639032</td>\n",
       "      <td>517.330541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>377.663295</td>\n",
       "      <td>441.179050</td>\n",
       "      <td>340.577492</td>\n",
       "      <td>580.087846</td>\n",
       "      <td>427.563337</td>\n",
       "      <td>470.395311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>1397.090092</td>\n",
       "      <td>1238.939929</td>\n",
       "      <td>1295.102678</td>\n",
       "      <td>1466.839507</td>\n",
       "      <td>1521.412874</td>\n",
       "      <td>1043.005123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16095 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_title         AF25         AF26         AF29      AF25DEX      AF26DEX  \\\n",
       "Gene_Name                                                                       \n",
       "A1BG-AS1       101.238412    69.133212    85.371425    75.057577    76.011260   \n",
       "A2M            947.239407   974.232499  1287.837024   792.393564   698.353451   \n",
       "A2M-AS1          8.803340    24.560483     5.449240     3.216753    14.252111   \n",
       "AAAS           572.217114   402.973854   455.011530   494.307758   597.400996   \n",
       "AACS           492.106718   352.033592   316.055913   551.137066   477.445726   \n",
       "...                   ...          ...          ...          ...          ...   \n",
       "ZXDA            47.538037    57.307794    35.420059    28.950780    45.131686   \n",
       "ZXDB            85.392400   140.085719   103.535558    60.046062    67.697528   \n",
       "ZXDC           450.731019   448.456230   479.533109   572.582088   516.639032   \n",
       "ZYG11B         377.663295   441.179050   340.577492   580.087846   427.563337   \n",
       "ZZEF1         1397.090092  1238.939929  1295.102678  1466.839507  1521.412874   \n",
       "\n",
       "Sample_title      AF29DEX  \n",
       "Gene_Name                  \n",
       "A1BG-AS1        57.365282  \n",
       "A2M            510.029505  \n",
       "A2M-AS1          0.000000  \n",
       "AAAS           493.341423  \n",
       "AACS           398.427957  \n",
       "...                   ...  \n",
       "ZXDA            32.333159  \n",
       "ZXDB            61.537302  \n",
       "ZXDC           517.330541  \n",
       "ZYG11B         470.395311  \n",
       "ZZEF1         1043.005123  \n",
       "\n",
       "[16095 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "216abac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE186950.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "368c3228",
   "metadata": {},
   "source": [
    "# GSE189182"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2e423164",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE189182_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE189182_readcount_in_vivo.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('gene_name')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8c85e504",
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_df = expr_df.iloc[21: , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "481fea4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE189182.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f418313",
   "metadata": {},
   "source": [
    "# GSE189305\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6ec001f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE189305_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE189305_readcount_in_vitro.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('gene_name')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "dbde8d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_df = expr_df.iloc[19: , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f0c27872",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE189305.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29a3501f",
   "metadata": {},
   "source": [
    "# GSE195617"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "efa90954",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5841784</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>1569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210194</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>5022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210195</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210196</th>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57773 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM5841784\n",
       "gene_id                    \n",
       "ENSG00000223972           0\n",
       "ENSG00000227232         892\n",
       "ENSG00000243485           0\n",
       "ENSG00000237613           0\n",
       "ENSG00000268020           0\n",
       "...                     ...\n",
       "ENSG00000198695        1569\n",
       "ENSG00000210194           7\n",
       "ENSG00000198727        5022\n",
       "ENSG00000210195           0\n",
       "ENSG00000210196          69\n",
       "\n",
       "[57773 rows x 1 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_counts = pd.read_csv('GSM5841784_Controlrawcounts.txt', sep=\" \", header=None).sort_index()\n",
    "control_counts = control_counts.rename(columns={0:'gene_id', 1:'GSM5841784'})\n",
    "control_counts = control_counts.set_index('gene_id')\n",
    "control_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "4a8eaf4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5841787</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>1103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210194</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>2843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210195</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210196</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57773 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM5841787\n",
       "gene_id                    \n",
       "ENSG00000223972           0\n",
       "ENSG00000227232         762\n",
       "ENSG00000243485           0\n",
       "ENSG00000237613           0\n",
       "ENSG00000268020           0\n",
       "...                     ...\n",
       "ENSG00000198695        1103\n",
       "ENSG00000210194           5\n",
       "ENSG00000198727        2843\n",
       "ENSG00000210195           0\n",
       "ENSG00000210196          68\n",
       "\n",
       "[57773 rows x 1 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dex_counts = pd.read_csv('GSM5841787_Dexamethasone_500uMrawcounts.txt', sep=\" \", header=None).sort_index()\n",
    "dex_counts = dex_counts.rename(columns={0:'gene_id', 1:'GSM5841787'})\n",
    "dex_counts = dex_counts.set_index('gene_id')\n",
    "dex_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "7d0087c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE195617_metadata.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "expr_df = pd.concat([control_counts, dex_counts], axis=1)\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "28453999",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute signature\n",
    "signature = logFC(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE195617.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca20297d",
   "metadata": {},
   "source": [
    "# GSE176277"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5384f0ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control mut'\n",
    "treatment = 'dex mut'\n",
    "meta_data_filename = 'GSE176277_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE176277_counts.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9cc7b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# signatures for mutated mice\n",
    "control_name = 'control mut'\n",
    "treatment = 'dex mut'\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE176277_mut.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "501334ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# signatures for wildtype mice\n",
    "control_name = 'control wt'\n",
    "treatment = 'dex wt'\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE176277_wt.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5de3aaa3",
   "metadata": {},
   "source": [
    "# GSE159952"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0eaa1af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control'\n",
    "treatment = 'dex'\n",
    "meta_data_filename = 'GSE159952_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE159952_all.tpm.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c678335",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE159952.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "135dc7e1",
   "metadata": {},
   "source": [
    "# GSE151572"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f936b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Treatment'\n",
    "control_name = 'control 24h'\n",
    "treatment = 'dex 24h'\n",
    "meta_data_filename = 'GSE151572_metadata.txt'\n",
    "rnaseq_data_filename = 'GSE151572_expressed_gene_reads.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db136898",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Treatment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_title</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Con_24h_1</th>\n",
       "      <td>control 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_24h_2</th>\n",
       "      <td>control 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_24h_3</th>\n",
       "      <td>control 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_6h_1</th>\n",
       "      <td>control 6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_6h_2</th>\n",
       "      <td>control 6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Con_6h_3</th>\n",
       "      <td>control 6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_24h_1</th>\n",
       "      <td>dex 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_24h_2</th>\n",
       "      <td>dex 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_24h_3</th>\n",
       "      <td>dex 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_6h_1</th>\n",
       "      <td>dex 6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_6h_2</th>\n",
       "      <td>dex 6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dex_6h_3</th>\n",
       "      <td>dex 6h</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Treatment\n",
       "Sample_title             \n",
       "Con_24h_1     control 24h\n",
       "Con_24h_2     control 24h\n",
       "Con_24h_3     control 24h\n",
       "Con_6h_1       control 6h\n",
       "Con_6h_2       control 6h\n",
       "Con_6h_3       control 6h\n",
       "Dex_24h_1         dex 24h\n",
       "Dex_24h_2         dex 24h\n",
       "Dex_24h_3         dex 24h\n",
       "Dex_6h_1           dex 6h\n",
       "Dex_6h_2           dex 6h\n",
       "Dex_6h_3           dex 6h"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "617b2bba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_title</th>\n",
       "      <th>Con_24h_1</th>\n",
       "      <th>Con_24h_2</th>\n",
       "      <th>Con_24h_3</th>\n",
       "      <th>Con_6h_1</th>\n",
       "      <th>Con_6h_2</th>\n",
       "      <th>Con_6h_3</th>\n",
       "      <th>Dex_24h_1</th>\n",
       "      <th>Dex_24h_2</th>\n",
       "      <th>Dex_24h_3</th>\n",
       "      <th>Dex_6h_1</th>\n",
       "      <th>Dex_6h_2</th>\n",
       "      <th>Dex_6h_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000000001</th>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>20</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000000007</th>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "      <td>37</td>\n",
       "      <td>10</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000000009</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000000010</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000000012</th>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>37</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000061925</th>\n",
       "      <td>44</td>\n",
       "      <td>15</td>\n",
       "      <td>25</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>16</td>\n",
       "      <td>71</td>\n",
       "      <td>34</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000061928</th>\n",
       "      <td>1065</td>\n",
       "      <td>1422</td>\n",
       "      <td>1373</td>\n",
       "      <td>1371</td>\n",
       "      <td>1356</td>\n",
       "      <td>1430</td>\n",
       "      <td>1251</td>\n",
       "      <td>1139</td>\n",
       "      <td>1494</td>\n",
       "      <td>1321</td>\n",
       "      <td>1193</td>\n",
       "      <td>1810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000061934</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOT00000072207</th>\n",
       "      <td>383</td>\n",
       "      <td>65</td>\n",
       "      <td>31</td>\n",
       "      <td>82</td>\n",
       "      <td>141</td>\n",
       "      <td>70</td>\n",
       "      <td>175</td>\n",
       "      <td>99</td>\n",
       "      <td>89</td>\n",
       "      <td>505</td>\n",
       "      <td>85</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOT00000076179</th>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>24</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>21</td>\n",
       "      <td>39</td>\n",
       "      <td>27</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19407 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_title        Con_24h_1  Con_24h_2  Con_24h_3  Con_6h_1  Con_6h_2  \\\n",
       "Gene                                                                      \n",
       "ENSRNOG00000000001         17         16          7         9        11   \n",
       "ENSRNOG00000000007          3          9         14        17        37   \n",
       "ENSRNOG00000000009          0          2          2         0         0   \n",
       "ENSRNOG00000000010          2          6          4         0        22   \n",
       "ENSRNOG00000000012         13         10          4         9        12   \n",
       "...                       ...        ...        ...       ...       ...   \n",
       "ENSRNOG00000061925         44         15         25        16        21   \n",
       "ENSRNOG00000061928       1065       1422       1373      1371      1356   \n",
       "ENSRNOG00000061934          1          2          0         0         0   \n",
       "ENSRNOT00000072207        383         65         31        82       141   \n",
       "ENSRNOT00000076179         27         19         24        30        39   \n",
       "\n",
       "Sample_title        Con_6h_3  Dex_24h_1  Dex_24h_2  Dex_24h_3  Dex_6h_1  \\\n",
       "Gene                                                                      \n",
       "ENSRNOG00000000001         8         18         15          6        22   \n",
       "ENSRNOG00000000007        10         18         23         40         8   \n",
       "ENSRNOG00000000009         0          2          0          0         0   \n",
       "ENSRNOG00000000010         4          0          1          2         3   \n",
       "ENSRNOG00000000012        12         31          3         21        37   \n",
       "...                      ...        ...        ...        ...       ...   \n",
       "ENSRNOG00000061925        21         35         25         16        71   \n",
       "ENSRNOG00000061928      1430       1251       1139       1494      1321   \n",
       "ENSRNOG00000061934         1          0          3          1        13   \n",
       "ENSRNOT00000072207        70        175         99         89       505   \n",
       "ENSRNOT00000076179        21         39         27         30        31   \n",
       "\n",
       "Sample_title        Dex_6h_2  Dex_6h_3  \n",
       "Gene                                    \n",
       "ENSRNOG00000000001        20        24  \n",
       "ENSRNOG00000000007        31        28  \n",
       "ENSRNOG00000000009         0         0  \n",
       "ENSRNOG00000000010         3         8  \n",
       "ENSRNOG00000000012        14        19  \n",
       "...                      ...       ...  \n",
       "ENSRNOG00000061925        34        37  \n",
       "ENSRNOG00000061928      1193      1810  \n",
       "ENSRNOG00000061934         6         5  \n",
       "ENSRNOT00000072207        85       130  \n",
       "ENSRNOT00000076179        30        38  \n",
       "\n",
       "[19407 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "69668fa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 24 hour signatures\n",
    "control_name = 'control 24h'\n",
    "treatment = 'dex 24h'\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE151572_24h.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bf6b4b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6 hour signatures\n",
    "control_name = 'control 6h'\n",
    "treatment = 'dex 6h'\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes'].to_csv('GSE151572_6h.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70aa7e2",
   "metadata": {},
   "source": [
    "# GSE149113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "5025c98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'vehicle'\n",
    "treatment = 'dexamethasone'\n",
    "meta_data_filename = 'GSE149113_series_matrix.txt'\n",
    "rnaseq_data_filename = 'GSE149113_NormCount.xlsx'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_excel(rnaseq_data_filename).sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "expr_df = expr_df.set_index('Gene_Name')\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "63ded6cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RMRP</th>\n",
       "      <td>0.262771</td>\n",
       "      <td>-0.250352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FKBP5</th>\n",
       "      <td>0.222084</td>\n",
       "      <td>-0.202657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H1B</th>\n",
       "      <td>0.217820</td>\n",
       "      <td>-0.229088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-RNR2</th>\n",
       "      <td>0.203790</td>\n",
       "      <td>-0.272381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL7R</th>\n",
       "      <td>0.190625</td>\n",
       "      <td>-0.241421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTSW</th>\n",
       "      <td>-0.093907</td>\n",
       "      <td>-0.154090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD44</th>\n",
       "      <td>-0.096249</td>\n",
       "      <td>-0.126348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BTG2</th>\n",
       "      <td>-0.105324</td>\n",
       "      <td>-0.104364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL2RB</th>\n",
       "      <td>-0.132847</td>\n",
       "      <td>-0.241186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>-0.194144</td>\n",
       "      <td>-0.039362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16087 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CD-coefficient  Significance\n",
       "Gene_Name                              \n",
       "RMRP             0.262771     -0.250352\n",
       "FKBP5            0.222084     -0.202657\n",
       "HIST1H1B         0.217820     -0.229088\n",
       "MT-RNR2          0.203790     -0.272381\n",
       "IL7R             0.190625     -0.241421\n",
       "...                   ...           ...\n",
       "CTSW            -0.093907     -0.154090\n",
       "CD44            -0.096249     -0.126348\n",
       "BTG2            -0.105324     -0.104364\n",
       "IL2RB           -0.132847     -0.241186\n",
       "ACTB            -0.194144     -0.039362\n",
       "\n",
       "[16087 rows x 2 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "a13cef03",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE149113.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1f69b32",
   "metadata": {},
   "source": [
    "# GSE141967"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "5abf1aff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error! Please load a metadata file in txt or tsv format\n",
      "Error! Please load a RNA-seq expression file in txt or tsv format\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"None of ['Gene_Name'] are in the columns\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [85]\u001b[0m, in \u001b[0;36m<cell line: 22>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m meta_class_column_name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m meta_df\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m     21\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError! Column \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmeta_class_column_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m is not in metadata\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 22\u001b[0m \u001b[43mexpr_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mset_index\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGene_Name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     23\u001b[0m meta_df \u001b[38;5;241m=\u001b[39m meta_df[meta_df\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39misin(expr_df\u001b[38;5;241m.\u001b[39mcolumns)]\n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m# sort metadata by class labels; control first\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/core/frame.py:5494\u001b[0m, in \u001b[0;36mDataFrame.set_index\u001b[0;34m(self, keys, drop, append, inplace, verify_integrity)\u001b[0m\n\u001b[1;32m   5491\u001b[0m                 missing\u001b[38;5;241m.\u001b[39mappend(col)\n\u001b[1;32m   5493\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m missing:\n\u001b[0;32m-> 5494\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmissing\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m are in the columns\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   5496\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[1;32m   5497\u001b[0m     frame \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of ['Gene_Name'] are in the columns\""
     ]
    }
   ],
   "source": [
    "# load in data\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment condition: VEH'\n",
    "treatment = 'treatment condition: DEX'\n",
    "meta_data_filename = 'GSE141967_series_matrix'\n",
    "rnaseq_data_filename = 'GSE141967_wandler_murine_tall_fpkm'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "90e1bfeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GM26870</th>\n",
       "      <td>0.503758</td>\n",
       "      <td>-0.099836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GM15662</th>\n",
       "      <td>0.276006</td>\n",
       "      <td>-0.091508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU3B3</th>\n",
       "      <td>0.243052</td>\n",
       "      <td>-0.102627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU3B1</th>\n",
       "      <td>0.243052</td>\n",
       "      <td>-0.102651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RMRP</th>\n",
       "      <td>0.160266</td>\n",
       "      <td>-0.103765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU2-10</th>\n",
       "      <td>-0.095053</td>\n",
       "      <td>-0.102675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GM23971</th>\n",
       "      <td>-0.095053</td>\n",
       "      <td>-0.097549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU3B2</th>\n",
       "      <td>-0.105754</td>\n",
       "      <td>-0.102639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNU3B4</th>\n",
       "      <td>-0.105754</td>\n",
       "      <td>-0.102615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GM24497</th>\n",
       "      <td>-0.229330</td>\n",
       "      <td>-0.097802</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16175 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CD-coefficient  Significance\n",
       "GM26870        0.503758     -0.099836\n",
       "GM15662        0.276006     -0.091508\n",
       "RNU3B3         0.243052     -0.102627\n",
       "RNU3B1         0.243052     -0.102651\n",
       "RMRP           0.160266     -0.103765\n",
       "...                 ...           ...\n",
       "RNU2-10       -0.095053     -0.102675\n",
       "GM23971       -0.095053     -0.097549\n",
       "RNU3B2        -0.105754     -0.102639\n",
       "RNU3B4        -0.105754     -0.102615\n",
       "GM24497       -0.229330     -0.097802\n",
       "\n",
       "[16175 rows x 2 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a72c5bba",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE141967.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "167a6f3e",
   "metadata": {},
   "source": [
    "# GSE137535"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "084c4a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: No Drug'\n",
    "treatment = 'treatment: DEX'\n",
    "meta_data_filename = 'GSE137535_series_matrix.txt'\n",
    "rnaseq_data_filename = 'GSE137535_MB231_DEXwSH454TreamtentsFeatureCountsedited_SH454wDEX.counts.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "f9803fc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM4081451</th>\n",
       "      <td>treatment: No Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081452</th>\n",
       "      <td>treatment: No Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081453</th>\n",
       "      <td>treatment: No Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081442</th>\n",
       "      <td>treatment: SH5-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081443</th>\n",
       "      <td>treatment: SH5-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081444</th>\n",
       "      <td>treatment: SH5-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081445</th>\n",
       "      <td>treatment: DEX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081446</th>\n",
       "      <td>treatment: DEX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081447</th>\n",
       "      <td>treatment: DEX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081448</th>\n",
       "      <td>treatment: DEX+SH4-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081449</th>\n",
       "      <td>treatment: DEX+SH4-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4081450</th>\n",
       "      <td>treatment: DEX+SH4-54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Sample_characteristics_ch1\n",
       "Sample_geo_accession                           \n",
       "GSM4081451                   treatment: No Drug\n",
       "GSM4081452                   treatment: No Drug\n",
       "GSM4081453                   treatment: No Drug\n",
       "GSM4081442                    treatment: SH5-54\n",
       "GSM4081443                    treatment: SH5-54\n",
       "GSM4081444                    treatment: SH5-54\n",
       "GSM4081445                       treatment: DEX\n",
       "GSM4081446                       treatment: DEX\n",
       "GSM4081447                       treatment: DEX\n",
       "GSM4081448                treatment: DEX+SH4-54\n",
       "GSM4081449                treatment: DEX+SH4-54\n",
       "GSM4081450                treatment: DEX+SH4-54"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b0651619",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>THBS1</th>\n",
       "      <td>0.689267</td>\n",
       "      <td>-0.000284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINE1</th>\n",
       "      <td>0.231069</td>\n",
       "      <td>-0.015688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANKRD1</th>\n",
       "      <td>0.182091</td>\n",
       "      <td>0.023851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANXA2</th>\n",
       "      <td>0.116774</td>\n",
       "      <td>0.025859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>0.112534</td>\n",
       "      <td>0.009169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIM</th>\n",
       "      <td>-0.070506</td>\n",
       "      <td>0.001158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKM</th>\n",
       "      <td>-0.086295</td>\n",
       "      <td>-0.036769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITGB1</th>\n",
       "      <td>-0.126496</td>\n",
       "      <td>-0.033229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FTL</th>\n",
       "      <td>-0.142021</td>\n",
       "      <td>-0.003442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEF1A1</th>\n",
       "      <td>-0.163795</td>\n",
       "      <td>0.017077</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19475 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CD-coefficient  Significance\n",
       "Geneid                                \n",
       "THBS1           0.689267     -0.000284\n",
       "SERPINE1        0.231069     -0.015688\n",
       "ANKRD1          0.182091      0.023851\n",
       "ANXA2           0.116774      0.025859\n",
       "ACTB            0.112534      0.009169\n",
       "...                  ...           ...\n",
       "VIM            -0.070506      0.001158\n",
       "PKM            -0.086295     -0.036769\n",
       "ITGB1          -0.126496     -0.033229\n",
       "FTL            -0.142021     -0.003442\n",
       "EEF1A1         -0.163795      0.017077\n",
       "\n",
       "[19475 rows x 2 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "70b65407",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE137535.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff87261f",
   "metadata": {},
   "source": [
    "# GSE119092"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bb8a08b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: control siRNA'\n",
    "treatment = 'treatment: control siRNA + DEX'\n",
    "meta_data_filename = 'GSE119092_series_matrix.txt'\n",
    "rnaseq_data_filename = 'GSE119092_gene-fpkm-table.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2151ae44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3357729</th>\n",
       "      <td>control siRNA [RNA-seq]</td>\n",
       "      <td>treatment: control siRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3357730</th>\n",
       "      <td>control siRNA + DEX [RNA-seq]</td>\n",
       "      <td>treatment: control siRNA + DEX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3357731</th>\n",
       "      <td>GR siRNA [RNA-seq]</td>\n",
       "      <td>treatment: GR siRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3357732</th>\n",
       "      <td>GR siRNA + DEX [RNA-seq]</td>\n",
       "      <td>treatment: GR siRNA + DEX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Sample_title  \\\n",
       "Sample_geo_accession                                  \n",
       "GSM3357729                  control siRNA [RNA-seq]   \n",
       "GSM3357730            control siRNA + DEX [RNA-seq]   \n",
       "GSM3357731                       GR siRNA [RNA-seq]   \n",
       "GSM3357732                 GR siRNA + DEX [RNA-seq]   \n",
       "\n",
       "                          Sample_characteristics_ch1  \n",
       "Sample_geo_accession                                  \n",
       "GSM3357729                  treatment: control siRNA  \n",
       "GSM3357730            treatment: control siRNA + DEX  \n",
       "GSM3357731                       treatment: GR siRNA  \n",
       "GSM3357732                 treatment: GR siRNA + DEX  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "830622dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM3357729</th>\n",
       "      <th>GSM3357730</th>\n",
       "      <th>GSM3357731</th>\n",
       "      <th>GSM3357732</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>34.51470</td>\n",
       "      <td>36.129600</td>\n",
       "      <td>31.53560</td>\n",
       "      <td>31.204600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>18.61930</td>\n",
       "      <td>18.948000</td>\n",
       "      <td>17.74630</td>\n",
       "      <td>22.355400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>1.12781</td>\n",
       "      <td>0.917989</td>\n",
       "      <td>1.30297</td>\n",
       "      <td>0.984707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>27.28910</td>\n",
       "      <td>25.834000</td>\n",
       "      <td>23.02970</td>\n",
       "      <td>22.621500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>7.39597</td>\n",
       "      <td>7.555420</td>\n",
       "      <td>7.45545</td>\n",
       "      <td>7.611530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zyg11b</th>\n",
       "      <td>6.92010</td>\n",
       "      <td>6.446620</td>\n",
       "      <td>6.46096</td>\n",
       "      <td>6.257910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zyx</th>\n",
       "      <td>158.55200</td>\n",
       "      <td>154.454000</td>\n",
       "      <td>186.03500</td>\n",
       "      <td>172.987000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzef1</th>\n",
       "      <td>8.35275</td>\n",
       "      <td>7.479040</td>\n",
       "      <td>9.22757</td>\n",
       "      <td>7.883160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzz3</th>\n",
       "      <td>12.25630</td>\n",
       "      <td>11.795000</td>\n",
       "      <td>10.57260</td>\n",
       "      <td>10.678900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>24.32730</td>\n",
       "      <td>25.626200</td>\n",
       "      <td>25.36490</td>\n",
       "      <td>25.286000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12199 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM3357729  GSM3357730  GSM3357731  GSM3357732\n",
       "0610007P14Rik           34.51470   36.129600    31.53560   31.204600\n",
       "0610009B22Rik           18.61930   18.948000    17.74630   22.355400\n",
       "0610009L18Rik            1.12781    0.917989     1.30297    0.984707\n",
       "0610009O20Rik           27.28910   25.834000    23.02970   22.621500\n",
       "0610010F05Rik            7.39597    7.555420     7.45545    7.611530\n",
       "...                          ...         ...         ...         ...\n",
       "Zyg11b                   6.92010    6.446620     6.46096    6.257910\n",
       "Zyx                    158.55200  154.454000   186.03500  172.987000\n",
       "Zzef1                    8.35275    7.479040     9.22757    7.883160\n",
       "Zzz3                    12.25630   11.795000    10.57260   10.678900\n",
       "l7Rn6                   24.32730   25.626200    25.36490   25.286000\n",
       "\n",
       "[12199 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3a96aa9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Apoo-ps</th>\n",
       "      <td>4.892671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mir6403</th>\n",
       "      <td>2.650563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Penk</th>\n",
       "      <td>2.338619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Defb25</th>\n",
       "      <td>2.287987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sap25</th>\n",
       "      <td>2.247801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mir6363</th>\n",
       "      <td>-1.841295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hist1h4a</th>\n",
       "      <td>-2.104328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1810012K16Rik</th>\n",
       "      <td>-2.326717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tmem190</th>\n",
       "      <td>-2.445362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hist1h2bm</th>\n",
       "      <td>-2.841292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12181 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  logFC\n",
       "Apoo-ps        4.892671\n",
       "Mir6403        2.650563\n",
       "Penk           2.338619\n",
       "Defb25         2.287987\n",
       "Sap25          2.247801\n",
       "...                 ...\n",
       "Mir6363       -1.841295\n",
       "Hist1h4a      -2.104328\n",
       "1810012K16Rik -2.326717\n",
       "Tmem190       -2.445362\n",
       "Hist1h2bm     -2.841292\n",
       "\n",
       "[12181 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures for dex pertubation\n",
    "signature = logFC(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bedfa4a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE119092_dex.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f5f401fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Apoo-ps</th>\n",
       "      <td>5.924251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rps27rt</th>\n",
       "      <td>5.148558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Defb25</th>\n",
       "      <td>3.124452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Penk</th>\n",
       "      <td>2.781566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arxes2</th>\n",
       "      <td>2.682746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lcn2</th>\n",
       "      <td>-3.274058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rmrp</th>\n",
       "      <td>-3.276428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hist2h4</th>\n",
       "      <td>-3.575981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mir546</th>\n",
       "      <td>-3.75073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rps27</th>\n",
       "      <td>-4.971482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12184 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC\n",
       "Apoo-ps  5.924251\n",
       "Rps27rt  5.148558\n",
       "Defb25   3.124452\n",
       "Penk     2.781566\n",
       "Arxes2   2.682746\n",
       "...           ...\n",
       "Lcn2    -3.274058\n",
       "Rmrp    -3.276428\n",
       "Hist2h4 -3.575981\n",
       "Mir546   -3.75073\n",
       "Rps27   -4.971482\n",
       "\n",
       "[12184 rows x 1 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# signatures for GR KO\n",
    "control_name = 'treatment: control siRNA'\n",
    "treatment = 'treatment: GR siRNA'\n",
    "signature = logFC(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "614830a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE119092_gr.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa4219b4",
   "metadata": {},
   "source": [
    "# GSE186104"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "18a40533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: standard cultivation medium'\n",
    "treatment = 'treatment: standard cultivation medium + dexamethason'\n",
    "meta_data_filename = 'GSE186104_series_matrix.txt'\n",
    "rnaseq_data_filename = 'GSE186104_cross_tabulation_of_gene_expression.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cbaca4e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5632354</th>\n",
       "      <td>C1: Untreated control replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632355</th>\n",
       "      <td>C2: Untreated control replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632356</th>\n",
       "      <td>C3: Untreated control replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632357</th>\n",
       "      <td>Dex1: Dexamethason treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632358</th>\n",
       "      <td>Dex2: Dexamethason treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632359</th>\n",
       "      <td>Dex3: Dexamethason treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       Sample_title  \\\n",
       "Sample_geo_accession                                                  \n",
       "GSM5632354                        C1: Untreated control replicate 1   \n",
       "GSM5632355                        C2: Untreated control replicate 2   \n",
       "GSM5632356                        C3: Untreated control replicate 3   \n",
       "GSM5632357            Dex1: Dexamethason treated sample replicate 1   \n",
       "GSM5632358            Dex2: Dexamethason treated sample replicate 2   \n",
       "GSM5632359            Dex3: Dexamethason treated sample replicate 3   \n",
       "\n",
       "                                             Sample_characteristics_ch1  \n",
       "Sample_geo_accession                                                     \n",
       "GSM5632354                       treatment: standard cultivation medium  \n",
       "GSM5632355                       treatment: standard cultivation medium  \n",
       "GSM5632356                       treatment: standard cultivation medium  \n",
       "GSM5632357            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632358            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632359            treatment: standard cultivation medium + dexam...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "db842927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>160</td>\n",
       "      <td>60</td>\n",
       "      <td>196</td>\n",
       "      <td>328</td>\n",
       "      <td>274</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>32</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>62</td>\n",
       "      <td>65</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>80</td>\n",
       "      <td>36</td>\n",
       "      <td>75</td>\n",
       "      <td>68</td>\n",
       "      <td>70</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aaas</th>\n",
       "      <td>541</td>\n",
       "      <td>306</td>\n",
       "      <td>506</td>\n",
       "      <td>295</td>\n",
       "      <td>298</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>392</td>\n",
       "      <td>199</td>\n",
       "      <td>385</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>374</td>\n",
       "      <td>218</td>\n",
       "      <td>233</td>\n",
       "      <td>144</td>\n",
       "      <td>192</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>611</td>\n",
       "      <td>344</td>\n",
       "      <td>560</td>\n",
       "      <td>222</td>\n",
       "      <td>316</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>1048</td>\n",
       "      <td>598</td>\n",
       "      <td>922</td>\n",
       "      <td>505</td>\n",
       "      <td>596</td>\n",
       "      <td>399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>987</td>\n",
       "      <td>453</td>\n",
       "      <td>863</td>\n",
       "      <td>453</td>\n",
       "      <td>705</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19641 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                         160          60         196         328   \n",
       "A2m                           32          15          19          62   \n",
       "A3galt2                       80          36          75          68   \n",
       "A4galt                         0           0           2           0   \n",
       "Aaas                         541         306         506         295   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        392         199         385         174   \n",
       "mrpl11                       374         218         233         144   \n",
       "mrpl24                       611         344         560         222   \n",
       "mrpl9                       1048         598         922         505   \n",
       "rnf141                       987         453         863         453   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  \n",
       "gene                                          \n",
       "A1i3                         274         287  \n",
       "A2m                           65          36  \n",
       "A3galt2                       70          47  \n",
       "A4galt                         1           2  \n",
       "Aaas                         298         209  \n",
       "...                          ...         ...  \n",
       "l7Rn6                        185         142  \n",
       "mrpl11                       192         133  \n",
       "mrpl24                       316         222  \n",
       "mrpl9                        596         399  \n",
       "rnf141                       705         387  \n",
       "\n",
       "[19641 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c0c5cf74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H19</th>\n",
       "      <td>0.540918</td>\n",
       "      <td>0.194621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col6a3</th>\n",
       "      <td>0.355815</td>\n",
       "      <td>0.127919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acan</th>\n",
       "      <td>0.233381</td>\n",
       "      <td>0.012220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Igf2</th>\n",
       "      <td>0.148825</td>\n",
       "      <td>0.197593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col2a1</th>\n",
       "      <td>0.134666</td>\n",
       "      <td>0.127396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serpinh1</th>\n",
       "      <td>-0.111309</td>\n",
       "      <td>0.056964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RGD1566401</th>\n",
       "      <td>-0.116253</td>\n",
       "      <td>0.076178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col9a1</th>\n",
       "      <td>-0.131756</td>\n",
       "      <td>0.128181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col27a1</th>\n",
       "      <td>-0.136371</td>\n",
       "      <td>0.127321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col11a1</th>\n",
       "      <td>-0.441890</td>\n",
       "      <td>0.126614</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19641 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CD-coefficient  Significance\n",
       "gene                                    \n",
       "H19               0.540918      0.194621\n",
       "Col6a3            0.355815      0.127919\n",
       "Acan              0.233381      0.012220\n",
       "Igf2              0.148825      0.197593\n",
       "Col2a1            0.134666      0.127396\n",
       "...                    ...           ...\n",
       "Serpinh1         -0.111309      0.056964\n",
       "RGD1566401       -0.116253      0.076178\n",
       "Col9a1           -0.131756      0.128181\n",
       "Col27a1          -0.136371      0.127321\n",
       "Col11a1          -0.441890      0.126614\n",
       "\n",
       "[19641 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14257400",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE186104.txt\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
