{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb498f7b",
   "metadata": {},
   "source": [
    "# Benchmarking RNA-seq DEG Methods with the Dexamethasone Benchmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f3fe067d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn import preprocessing\n",
    "import warnings\n",
    "import scipy.stats as ss\n",
    "from itertools import combinations\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcdbaf65",
   "metadata": {},
   "source": [
    "# Load in Data\n",
    "\n",
    "Using data from GEO from the study, \"The effect of lithium and dexamethasone on fetal rat metatarsal bones transcriptome\" \n",
    "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186104 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cc26d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "meta_df = pd.read_csv('GSE186104_series_matrix.txt', sep=\"\\t\", index_col=0, dtype=str)\n",
    "expr_df = pd.read_csv('GSE186104_cross_tabulation_of_gene_expression.txt', index_col=0, sep=\"\\t\").sort_index()\n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: standard cultivation medium'\n",
    "\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset[\"dataset_metadata\"] = meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36e3e6db",
   "metadata": {},
   "outputs": [],
   "source": [
    "low_expression_threshold = 0.3\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3603da18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5632354</th>\n",
       "      <td>C1: Untreated control replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632355</th>\n",
       "      <td>C2: Untreated control replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632356</th>\n",
       "      <td>C3: Untreated control replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632357</th>\n",
       "      <td>Dex1: Dexamethason treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632358</th>\n",
       "      <td>Dex2: Dexamethason treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632359</th>\n",
       "      <td>Dex3: Dexamethason treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632360</th>\n",
       "      <td>Li1: Lithium treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632361</th>\n",
       "      <td>Li2: Lithium treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632362</th>\n",
       "      <td>Li3: Lithium treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + lithium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632363</th>\n",
       "      <td>DL1: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632364</th>\n",
       "      <td>DL2: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632365</th>\n",
       "      <td>DL3: Dexamethason and lithium treated sample r...</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Sample_title  \\\n",
       "Sample_geo_accession                                                      \n",
       "GSM5632354                            C1: Untreated control replicate 1   \n",
       "GSM5632355                            C2: Untreated control replicate 2   \n",
       "GSM5632356                            C3: Untreated control replicate 3   \n",
       "GSM5632357                Dex1: Dexamethason treated sample replicate 1   \n",
       "GSM5632358                Dex2: Dexamethason treated sample replicate 2   \n",
       "GSM5632359                Dex3: Dexamethason treated sample replicate 3   \n",
       "GSM5632360                      Li1: Lithium treated sample replicate 1   \n",
       "GSM5632361                      Li2: Lithium treated sample replicate 2   \n",
       "GSM5632362                      Li3: Lithium treated sample replicate 3   \n",
       "GSM5632363            DL1: Dexamethason and lithium treated sample r...   \n",
       "GSM5632364            DL2: Dexamethason and lithium treated sample r...   \n",
       "GSM5632365            DL3: Dexamethason and lithium treated sample r...   \n",
       "\n",
       "                                             Sample_characteristics_ch1  \n",
       "Sample_geo_accession                                                     \n",
       "GSM5632354                       treatment: standard cultivation medium  \n",
       "GSM5632355                       treatment: standard cultivation medium  \n",
       "GSM5632356                       treatment: standard cultivation medium  \n",
       "GSM5632357            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632358            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632359            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632360             treatment: standard cultivation medium + lithium  \n",
       "GSM5632361             treatment: standard cultivation medium + lithium  \n",
       "GSM5632362             treatment: standard cultivation medium + lithium  \n",
       "GSM5632363            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632364            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632365            treatment: standard cultivation medium + dexam...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86d112cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "      <th>GSM5632360</th>\n",
       "      <th>GSM5632361</th>\n",
       "      <th>GSM5632362</th>\n",
       "      <th>GSM5632363</th>\n",
       "      <th>GSM5632364</th>\n",
       "      <th>GSM5632365</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>160</td>\n",
       "      <td>60</td>\n",
       "      <td>196</td>\n",
       "      <td>328</td>\n",
       "      <td>274</td>\n",
       "      <td>287</td>\n",
       "      <td>179</td>\n",
       "      <td>188</td>\n",
       "      <td>343</td>\n",
       "      <td>113</td>\n",
       "      <td>93</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>32</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>62</td>\n",
       "      <td>65</td>\n",
       "      <td>36</td>\n",
       "      <td>28</td>\n",
       "      <td>34</td>\n",
       "      <td>68</td>\n",
       "      <td>23</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ml1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>80</td>\n",
       "      <td>36</td>\n",
       "      <td>75</td>\n",
       "      <td>68</td>\n",
       "      <td>70</td>\n",
       "      <td>47</td>\n",
       "      <td>63</td>\n",
       "      <td>44</td>\n",
       "      <td>104</td>\n",
       "      <td>55</td>\n",
       "      <td>69</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>392</td>\n",
       "      <td>199</td>\n",
       "      <td>385</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "      <td>142</td>\n",
       "      <td>189</td>\n",
       "      <td>245</td>\n",
       "      <td>303</td>\n",
       "      <td>269</td>\n",
       "      <td>361</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>374</td>\n",
       "      <td>218</td>\n",
       "      <td>233</td>\n",
       "      <td>144</td>\n",
       "      <td>192</td>\n",
       "      <td>133</td>\n",
       "      <td>156</td>\n",
       "      <td>176</td>\n",
       "      <td>244</td>\n",
       "      <td>210</td>\n",
       "      <td>327</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>611</td>\n",
       "      <td>344</td>\n",
       "      <td>560</td>\n",
       "      <td>222</td>\n",
       "      <td>316</td>\n",
       "      <td>222</td>\n",
       "      <td>286</td>\n",
       "      <td>327</td>\n",
       "      <td>497</td>\n",
       "      <td>369</td>\n",
       "      <td>513</td>\n",
       "      <td>569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>1048</td>\n",
       "      <td>598</td>\n",
       "      <td>922</td>\n",
       "      <td>505</td>\n",
       "      <td>596</td>\n",
       "      <td>399</td>\n",
       "      <td>493</td>\n",
       "      <td>646</td>\n",
       "      <td>954</td>\n",
       "      <td>684</td>\n",
       "      <td>1049</td>\n",
       "      <td>1192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>987</td>\n",
       "      <td>453</td>\n",
       "      <td>863</td>\n",
       "      <td>453</td>\n",
       "      <td>705</td>\n",
       "      <td>387</td>\n",
       "      <td>450</td>\n",
       "      <td>609</td>\n",
       "      <td>798</td>\n",
       "      <td>691</td>\n",
       "      <td>1016</td>\n",
       "      <td>1014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21199 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                         160          60         196         328   \n",
       "A2m                           32          15          19          62   \n",
       "A2ml1                          1           0           0           0   \n",
       "A3galt2                       80          36          75          68   \n",
       "A4galt                         0           0           2           0   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        392         199         385         174   \n",
       "mrpl11                       374         218         233         144   \n",
       "mrpl24                       611         344         560         222   \n",
       "mrpl9                       1048         598         922         505   \n",
       "rnf141                       987         453         863         453   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  GSM5632360  GSM5632361  \\\n",
       "gene                                                                   \n",
       "A1i3                         274         287         179         188   \n",
       "A2m                           65          36          28          34   \n",
       "A2ml1                          2           0           3           0   \n",
       "A3galt2                       70          47          63          44   \n",
       "A4galt                         1           2           0           0   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        185         142         189         245   \n",
       "mrpl11                       192         133         156         176   \n",
       "mrpl24                       316         222         286         327   \n",
       "mrpl9                        596         399         493         646   \n",
       "rnf141                       705         387         450         609   \n",
       "\n",
       "Sample_geo_accession  GSM5632362  GSM5632363  GSM5632364  GSM5632365  \n",
       "gene                                                                  \n",
       "A1i3                         343         113          93         119  \n",
       "A2m                           68          23          29          29  \n",
       "A2ml1                          0           2           4           1  \n",
       "A3galt2                      104          55          69          76  \n",
       "A4galt                         0           0           1           0  \n",
       "...                          ...         ...         ...         ...  \n",
       "l7Rn6                        303         269         361         407  \n",
       "mrpl11                       244         210         327         287  \n",
       "mrpl24                       497         369         513         569  \n",
       "mrpl9                        954         684        1049        1192  \n",
       "rnf141                       798         691        1016        1014  \n",
       "\n",
       "[21199 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ff8e58ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def logCPM(data):\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        data = (data/data.sum())*10**6\n",
    "        data = data.fillna(0)\n",
    "        data = np.log2(data+1)\n",
    "\n",
    "    # Return\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5a843c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize(dataset, current_dataset, logCPM_normalization, log_normalization, z_normalization, q_normalization):\n",
    "    normalization = current_dataset\n",
    "    if logCPM_normalization == True:  \n",
    "        data = dataset[normalization]\n",
    "        normalization += '+logCPM'\n",
    "        dataset[normalization] = logCPM(data)\n",
    "        \n",
    "    if log_normalization == True:    \n",
    "        data = dataset[normalization]\n",
    "        normalization += '+log'\n",
    "        dataset[normalization] = log(data)\n",
    "        \n",
    "    if z_normalization == True:\n",
    "        data = dataset[normalization]\n",
    "        normalization += '+z_norm'    \n",
    "        dataset[normalization] = data.T.apply(ss.zscore, axis=0).T.dropna()\n",
    "\n",
    "    if q_normalization == True:\n",
    "        data = dataset[normalization]\n",
    "        normalization += '+q_norm'\n",
    "        dataset[normalization] = qnormalization(data)\n",
    "    return dataset, normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e5218a4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dataset, normalization = normalize(dataset, current_dataset, True, False, True, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bf46709c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "      <th>GSM5632360</th>\n",
       "      <th>GSM5632361</th>\n",
       "      <th>GSM5632362</th>\n",
       "      <th>GSM5632363</th>\n",
       "      <th>GSM5632364</th>\n",
       "      <th>GSM5632365</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>-0.823915</td>\n",
       "      <td>-1.203701</td>\n",
       "      <td>-0.394838</td>\n",
       "      <td>1.472828</td>\n",
       "      <td>0.950546</td>\n",
       "      <td>1.391266</td>\n",
       "      <td>0.682852</td>\n",
       "      <td>0.464780</td>\n",
       "      <td>0.636785</td>\n",
       "      <td>-0.661800</td>\n",
       "      <td>-1.357623</td>\n",
       "      <td>-1.157182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>-0.850702</td>\n",
       "      <td>-0.913618</td>\n",
       "      <td>-1.407946</td>\n",
       "      <td>1.707599</td>\n",
       "      <td>1.497647</td>\n",
       "      <td>0.792520</td>\n",
       "      <td>0.368323</td>\n",
       "      <td>0.377260</td>\n",
       "      <td>0.752250</td>\n",
       "      <td>-0.662710</td>\n",
       "      <td>-0.757919</td>\n",
       "      <td>-0.902703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ml1</th>\n",
       "      <td>-0.294967</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>0.920040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>2.292632</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>-0.772040</td>\n",
       "      <td>0.670972</td>\n",
       "      <td>1.308982</td>\n",
       "      <td>-0.265415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>-0.875900</td>\n",
       "      <td>-1.189250</td>\n",
       "      <td>-0.782604</td>\n",
       "      <td>1.542712</td>\n",
       "      <td>1.049973</td>\n",
       "      <td>0.388160</td>\n",
       "      <td>1.659979</td>\n",
       "      <td>-0.508135</td>\n",
       "      <td>0.933154</td>\n",
       "      <td>-0.559917</td>\n",
       "      <td>-0.820001</td>\n",
       "      <td>-0.838169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>1.011257</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>0.792626</td>\n",
       "      <td>2.716666</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>-0.600760</td>\n",
       "      <td>0.285531</td>\n",
       "      <td>-0.600760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>-0.314866</td>\n",
       "      <td>0.016371</td>\n",
       "      <td>0.365216</td>\n",
       "      <td>-0.613465</td>\n",
       "      <td>-1.450318</td>\n",
       "      <td>-1.753019</td>\n",
       "      <td>1.168210</td>\n",
       "      <td>1.839965</td>\n",
       "      <td>-0.805107</td>\n",
       "      <td>0.471590</td>\n",
       "      <td>0.450185</td>\n",
       "      <td>0.625238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>0.682765</td>\n",
       "      <td>1.931779</td>\n",
       "      <td>-2.038182</td>\n",
       "      <td>-0.558921</td>\n",
       "      <td>0.423214</td>\n",
       "      <td>-0.539152</td>\n",
       "      <td>0.762820</td>\n",
       "      <td>0.270977</td>\n",
       "      <td>-0.896703</td>\n",
       "      <td>-0.159605</td>\n",
       "      <td>0.889206</td>\n",
       "      <td>-0.768198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>0.115012</td>\n",
       "      <td>1.517950</td>\n",
       "      <td>0.140863</td>\n",
       "      <td>-2.218022</td>\n",
       "      <td>-0.175757</td>\n",
       "      <td>-1.392579</td>\n",
       "      <td>1.395951</td>\n",
       "      <td>0.835633</td>\n",
       "      <td>0.106610</td>\n",
       "      <td>-0.338141</td>\n",
       "      <td>-0.004243</td>\n",
       "      <td>0.016723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>-0.874443</td>\n",
       "      <td>0.767789</td>\n",
       "      <td>-1.288662</td>\n",
       "      <td>-0.328218</td>\n",
       "      <td>-0.159364</td>\n",
       "      <td>-1.983152</td>\n",
       "      <td>0.546399</td>\n",
       "      <td>1.424708</td>\n",
       "      <td>0.333021</td>\n",
       "      <td>-0.520593</td>\n",
       "      <td>0.898979</td>\n",
       "      <td>1.183536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>-0.587765</td>\n",
       "      <td>-1.164860</td>\n",
       "      <td>-0.978082</td>\n",
       "      <td>-0.571429</td>\n",
       "      <td>1.998474</td>\n",
       "      <td>-1.230443</td>\n",
       "      <td>0.291034</td>\n",
       "      <td>1.288092</td>\n",
       "      <td>-0.648133</td>\n",
       "      <td>0.315527</td>\n",
       "      <td>1.098828</td>\n",
       "      <td>0.188757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21199 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                   -0.823915   -1.203701   -0.394838    1.472828   \n",
       "A2m                    -0.850702   -0.913618   -1.407946    1.707599   \n",
       "A2ml1                  -0.294967   -0.772040   -0.772040   -0.772040   \n",
       "A3galt2                -0.875900   -1.189250   -0.782604    1.542712   \n",
       "A4galt                 -0.600760   -0.600760    1.011257   -0.600760   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                  -0.314866    0.016371    0.365216   -0.613465   \n",
       "mrpl11                  0.682765    1.931779   -2.038182   -0.558921   \n",
       "mrpl24                  0.115012    1.517950    0.140863   -2.218022   \n",
       "mrpl9                  -0.874443    0.767789   -1.288662   -0.328218   \n",
       "rnf141                 -0.587765   -1.164860   -0.978082   -0.571429   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  GSM5632360  GSM5632361  \\\n",
       "gene                                                                   \n",
       "A1i3                    0.950546    1.391266    0.682852    0.464780   \n",
       "A2m                     1.497647    0.792520    0.368323    0.377260   \n",
       "A2ml1                   0.920040   -0.772040    2.292632   -0.772040   \n",
       "A3galt2                 1.049973    0.388160    1.659979   -0.508135   \n",
       "A4galt                  0.792626    2.716666   -0.600760   -0.600760   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                  -1.450318   -1.753019    1.168210    1.839965   \n",
       "mrpl11                  0.423214   -0.539152    0.762820    0.270977   \n",
       "mrpl24                 -0.175757   -1.392579    1.395951    0.835633   \n",
       "mrpl9                  -0.159364   -1.983152    0.546399    1.424708   \n",
       "rnf141                  1.998474   -1.230443    0.291034    1.288092   \n",
       "\n",
       "Sample_geo_accession  GSM5632362  GSM5632363  GSM5632364  GSM5632365  \n",
       "gene                                                                  \n",
       "A1i3                    0.636785   -0.661800   -1.357623   -1.157182  \n",
       "A2m                     0.752250   -0.662710   -0.757919   -0.902703  \n",
       "A2ml1                  -0.772040    0.670972    1.308982   -0.265415  \n",
       "A3galt2                 0.933154   -0.559917   -0.820001   -0.838169  \n",
       "A4galt                 -0.600760   -0.600760    0.285531   -0.600760  \n",
       "...                          ...         ...         ...         ...  \n",
       "l7Rn6                  -0.805107    0.471590    0.450185    0.625238  \n",
       "mrpl11                 -0.896703   -0.159605    0.889206   -0.768198  \n",
       "mrpl24                  0.106610   -0.338141   -0.004243    0.016723  \n",
       "mrpl9                   0.333021   -0.520593    0.898979    1.183536  \n",
       "rnf141                 -0.648133    0.315527    1.098828    0.188757  \n",
       "\n",
       "[21199 rows x 12 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset[normalization]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc27b4b4",
   "metadata": {},
   "source": [
    "# Differential Gene Expression\n",
    "Using code adapted from Bulk RNA-seq Analysis pipeline appyter: https://appyters.maayanlab.cloud/Bulk_RNA_seq/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d9bcc6d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from the appyter source code\n",
    "def get_signatures(classes, dataset, normalization, method, meta_class_column_name, filter_genes):\n",
    "    tmp_normalization = normalization.replace(\"+z_norm+q_norm\",\"\").replace(\"+z_norm\",\"\")\n",
    "    raw_expr_df = dataset['rawdata']\n",
    "    expr_df = dataset['rawdata']\n",
    "    if filter_genes == True:\n",
    "        expr_df = dataset['rawdata+filter_genes']\n",
    "        \n",
    "    signatures = dict()\n",
    "\n",
    "    for cls1, cls2 in combinations(classes, 2):\n",
    "        print(cls1, cls2)\n",
    "        cls1_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==cls1, :].index.tolist() #control\n",
    "        cls2_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==cls2,:].index.tolist() #case\n",
    "        \n",
    "        signature_label = \" vs. \".join([cls1, cls2])\n",
    "        \n",
    "        if method == \"limma\":\n",
    "            limma = robjects.r['limma']\n",
    "\n",
    "            design_dataframe = pd.DataFrame([{'index': x, 'A': int(x in cls1_sample_ids), 'B': int(x in cls2_sample_ids)} for x in raw_expr_df.columns]).set_index('index')\n",
    "\n",
    "            processed_data = {\"expression\": raw_expr_df, 'design': design_dataframe}\n",
    "            \n",
    "            limma_results = pandas2ri.conversion.rpy2py(limma(pandas2ri.conversion.py2rpy(processed_data['expression']), pandas2ri.conversion.py2rpy(processed_data['design']), filter_genes=filter_genes))\n",
    "                        \n",
    "            signature = pd.DataFrame(limma_results[0])\n",
    "            signature.index = limma_results[1]\n",
    "            signature = signature.sort_values(\"t\", ascending=False)\n",
    "            \n",
    "        elif method == \"characteristic_direction\":\n",
    "            signature = characteristic_direction(dataset[tmp_normalization].loc[:, cls1_sample_ids], dataset[normalization].loc[:, cls2_sample_ids], calculate_sig=True)\n",
    "            signature = signature.sort_values(\"CD-coefficient\", ascending=False)\n",
    "            \n",
    "        signatures[signature_label] = signature\n",
    "\n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2036a595",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "treatment: standard cultivation medium treatment: standard cultivation medium + dexamethason\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"'limma' not found\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [15]\u001b[0m, in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Set method in this variable\u001b[39;00m\n\u001b[1;32m      2\u001b[0m diff_gex_method \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlimma\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 4\u001b[0m signatures \u001b[38;5;241m=\u001b[39m \u001b[43mget_signatures\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclasses\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnormalization\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdiff_gex_method\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmeta_class_column_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [10]\u001b[0m, in \u001b[0;36mget_signatures\u001b[0;34m(classes, dataset, normalization, method, meta_class_column_name, filter_genes)\u001b[0m\n\u001b[1;32m     16\u001b[0m signature_label \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m vs. \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin([cls1, cls2])\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlimma\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 19\u001b[0m     limma \u001b[38;5;241m=\u001b[39m \u001b[43mrobjects\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mr\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mlimma\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     21\u001b[0m     design_dataframe \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame([{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m: x, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28mint\u001b[39m(x \u001b[38;5;129;01min\u001b[39;00m cls1_sample_ids), \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28mint\u001b[39m(x \u001b[38;5;129;01min\u001b[39;00m cls2_sample_ids)} \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m raw_expr_df\u001b[38;5;241m.\u001b[39mcolumns])\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     23\u001b[0m     processed_data \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexpression\u001b[39m\u001b[38;5;124m\"\u001b[39m: raw_expr_df, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdesign\u001b[39m\u001b[38;5;124m'\u001b[39m: design_dataframe}\n",
      "File \u001b[0;32m~/Documents/Harvard/Summer 2022/Mount Sinai/venv3.9/lib/python3.9/site-packages/rpy2/robjects/__init__.py:439\u001b[0m, in \u001b[0;36mR.__getitem__\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, item):\n\u001b[0;32m--> 439\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[43m_globalenv\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfind\u001b[49m\u001b[43m(\u001b[49m\u001b[43mitem\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    440\u001b[0m     res \u001b[38;5;241m=\u001b[39m conversion\u001b[38;5;241m.\u001b[39mget_conversion()\u001b[38;5;241m.\u001b[39mrpy2py(res)\n\u001b[1;32m    441\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(res, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__rname__\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[0;32m~/Documents/Harvard/Summer 2022/Mount Sinai/venv3.9/lib/python3.9/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[38;5;241m=\u001b[39m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;66;03m# TODO: test cdata is of the expected CType\u001b[39;00m\n\u001b[1;32m     47\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _cdata_to_rinterface(cdata)\n",
      "File \u001b[0;32m~/Documents/Harvard/Summer 2022/Mount Sinai/venv3.9/lib/python3.9/site-packages/rpy2/rinterface_lib/_rinterface_capi.py:333\u001b[0m, in \u001b[0;36m_evaluated_promise.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    332\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m--> 333\u001b[0m     robj \u001b[38;5;241m=\u001b[39m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    334\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m _TYPEOF(robj) \u001b[38;5;241m==\u001b[39m openrlib\u001b[38;5;241m.\u001b[39mrlib\u001b[38;5;241m.\u001b[39mPROMSXP:\n\u001b[1;32m    335\u001b[0m         robj \u001b[38;5;241m=\u001b[39m openrlib\u001b[38;5;241m.\u001b[39mrlib\u001b[38;5;241m.\u001b[39mRf_eval(\n\u001b[1;32m    336\u001b[0m             robj,\n\u001b[1;32m    337\u001b[0m             openrlib\u001b[38;5;241m.\u001b[39mrlib\u001b[38;5;241m.\u001b[39mR_GlobalEnv)\n",
      "File \u001b[0;32m~/Documents/Harvard/Summer 2022/Mount Sinai/venv3.9/lib/python3.9/site-packages/rpy2/rinterface_lib/sexp.py:357\u001b[0m, in \u001b[0;36mSexpEnvironment.find\u001b[0;34m(self, key, wantfun)\u001b[0m\n\u001b[1;32m    355\u001b[0m \u001b[38;5;66;03m# TODO: move check of R_UnboundValue to _rinterface ?\u001b[39;00m\n\u001b[1;32m    356\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m res \u001b[38;5;241m==\u001b[39m openrlib\u001b[38;5;241m.\u001b[39mrlib\u001b[38;5;241m.\u001b[39mR_UnboundValue:\n\u001b[0;32m--> 357\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m not found\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m key)\n\u001b[1;32m    358\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "\u001b[0;31mKeyError\u001b[0m: \"'limma' not found\""
     ]
    }
   ],
   "source": [
    "# Set method in this variable\n",
    "diff_gex_method = 'limma'\n",
    "\n",
    "signatures = get_signatures(classes, dataset, normalization, diff_gex_method, meta_class_column_name, True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3abb034c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cf21f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract NR3C1 rankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49bd1080",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare rankings/methods"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv3.9",
   "language": "python",
   "name": "venv3.9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
