{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "de55cbab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import dependencies\n",
    "import pandas as pd\n",
    "from maayanlab_bioinformatics.dge.characteristic_direction import characteristic_direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c50adeac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for computing signatures with characteristic direction\n",
    "def cd_signature(control, treatment, dataset, normalization, meta_class_column_name):\n",
    "    tmp_normalization = normalization\n",
    "    \n",
    "    signatures = dict()\n",
    "    \n",
    "    signature_label = normalization\n",
    "    \n",
    "    cls1_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==control, :].index.tolist() #control\n",
    "    cls2_sample_ids = dataset[\"dataset_metadata\"].loc[dataset[\"dataset_metadata\"][meta_class_column_name]==treatment,:].index.tolist() #case\n",
    "    \n",
    "    signature = characteristic_direction(dataset[tmp_normalization].loc[:, cls1_sample_ids], dataset[normalization].loc[:, cls2_sample_ids], calculate_sig=True)\n",
    "    signature = signature.sort_values(\"CD-coefficient\", ascending=False)\n",
    "    \n",
    "    signatures[signature_label] = signature\n",
    "    \n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa4219b4",
   "metadata": {},
   "source": [
    "# GSE186104"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "18a40533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data \n",
    "meta_class_column_name = 'Sample_characteristics_ch1'\n",
    "control_name = 'treatment: standard cultivation medium'\n",
    "treatment = 'treatment: standard cultivation medium + dexamethason'\n",
    "meta_data_filename = 'GSE186104_series_matrix.txt'\n",
    "rnaseq_data_filename = 'GSE186104_cross_tabulation_of_gene_expression.txt'\n",
    "low_expression_threshold = 0.3\n",
    "\n",
    "try:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt or tsv format\")\n",
    "try:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "except:\n",
    "    print(\"Error! Please load a RNA-seq expression file in txt or tsv format\")\n",
    "meta_df.index = meta_df.index.map(str)\n",
    "\n",
    "# Match samples between the metadata and the datasets\n",
    "if meta_class_column_name not in meta_df.columns:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "    \n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format.\")\n",
    "\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cbaca4e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_title</th>\n",
       "      <th>Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5632354</th>\n",
       "      <td>C1: Untreated control replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632355</th>\n",
       "      <td>C2: Untreated control replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632356</th>\n",
       "      <td>C3: Untreated control replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632357</th>\n",
       "      <td>Dex1: Dexamethason treated sample replicate 1</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632358</th>\n",
       "      <td>Dex2: Dexamethason treated sample replicate 2</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5632359</th>\n",
       "      <td>Dex3: Dexamethason treated sample replicate 3</td>\n",
       "      <td>treatment: standard cultivation medium + dexam...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       Sample_title  \\\n",
       "Sample_geo_accession                                                  \n",
       "GSM5632354                        C1: Untreated control replicate 1   \n",
       "GSM5632355                        C2: Untreated control replicate 2   \n",
       "GSM5632356                        C3: Untreated control replicate 3   \n",
       "GSM5632357            Dex1: Dexamethason treated sample replicate 1   \n",
       "GSM5632358            Dex2: Dexamethason treated sample replicate 2   \n",
       "GSM5632359            Dex3: Dexamethason treated sample replicate 3   \n",
       "\n",
       "                                             Sample_characteristics_ch1  \n",
       "Sample_geo_accession                                                     \n",
       "GSM5632354                       treatment: standard cultivation medium  \n",
       "GSM5632355                       treatment: standard cultivation medium  \n",
       "GSM5632356                       treatment: standard cultivation medium  \n",
       "GSM5632357            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632358            treatment: standard cultivation medium + dexam...  \n",
       "GSM5632359            treatment: standard cultivation medium + dexam...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db842927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample_geo_accession</th>\n",
       "      <th>GSM5632354</th>\n",
       "      <th>GSM5632355</th>\n",
       "      <th>GSM5632356</th>\n",
       "      <th>GSM5632357</th>\n",
       "      <th>GSM5632358</th>\n",
       "      <th>GSM5632359</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1i3</th>\n",
       "      <td>160</td>\n",
       "      <td>60</td>\n",
       "      <td>196</td>\n",
       "      <td>328</td>\n",
       "      <td>274</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>32</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>62</td>\n",
       "      <td>65</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>80</td>\n",
       "      <td>36</td>\n",
       "      <td>75</td>\n",
       "      <td>68</td>\n",
       "      <td>70</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aaas</th>\n",
       "      <td>541</td>\n",
       "      <td>306</td>\n",
       "      <td>506</td>\n",
       "      <td>295</td>\n",
       "      <td>298</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>392</td>\n",
       "      <td>199</td>\n",
       "      <td>385</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl11</th>\n",
       "      <td>374</td>\n",
       "      <td>218</td>\n",
       "      <td>233</td>\n",
       "      <td>144</td>\n",
       "      <td>192</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl24</th>\n",
       "      <td>611</td>\n",
       "      <td>344</td>\n",
       "      <td>560</td>\n",
       "      <td>222</td>\n",
       "      <td>316</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mrpl9</th>\n",
       "      <td>1048</td>\n",
       "      <td>598</td>\n",
       "      <td>922</td>\n",
       "      <td>505</td>\n",
       "      <td>596</td>\n",
       "      <td>399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rnf141</th>\n",
       "      <td>987</td>\n",
       "      <td>453</td>\n",
       "      <td>863</td>\n",
       "      <td>453</td>\n",
       "      <td>705</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19641 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample_geo_accession  GSM5632354  GSM5632355  GSM5632356  GSM5632357  \\\n",
       "gene                                                                   \n",
       "A1i3                         160          60         196         328   \n",
       "A2m                           32          15          19          62   \n",
       "A3galt2                       80          36          75          68   \n",
       "A4galt                         0           0           2           0   \n",
       "Aaas                         541         306         506         295   \n",
       "...                          ...         ...         ...         ...   \n",
       "l7Rn6                        392         199         385         174   \n",
       "mrpl11                       374         218         233         144   \n",
       "mrpl24                       611         344         560         222   \n",
       "mrpl9                       1048         598         922         505   \n",
       "rnf141                       987         453         863         453   \n",
       "\n",
       "Sample_geo_accession  GSM5632358  GSM5632359  \n",
       "gene                                          \n",
       "A1i3                         274         287  \n",
       "A2m                           65          36  \n",
       "A3galt2                       70          47  \n",
       "A4galt                         1           2  \n",
       "Aaas                         298         209  \n",
       "...                          ...         ...  \n",
       "l7Rn6                        185         142  \n",
       "mrpl11                       192         133  \n",
       "mrpl24                       316         222  \n",
       "mrpl9                        596         399  \n",
       "rnf141                       705         387  \n",
       "\n",
       "[19641 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c0c5cf74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD-coefficient</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H19</th>\n",
       "      <td>0.540918</td>\n",
       "      <td>0.064525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col6a3</th>\n",
       "      <td>0.355815</td>\n",
       "      <td>0.055393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acan</th>\n",
       "      <td>0.233381</td>\n",
       "      <td>0.007798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Igf2</th>\n",
       "      <td>0.148825</td>\n",
       "      <td>0.061754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col2a1</th>\n",
       "      <td>0.134666</td>\n",
       "      <td>0.055251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serpinh1</th>\n",
       "      <td>-0.111309</td>\n",
       "      <td>0.011136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RGD1566401</th>\n",
       "      <td>-0.116253</td>\n",
       "      <td>0.012981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col9a1</th>\n",
       "      <td>-0.131756</td>\n",
       "      <td>0.055465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col27a1</th>\n",
       "      <td>-0.136371</td>\n",
       "      <td>0.055230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Col11a1</th>\n",
       "      <td>-0.441890</td>\n",
       "      <td>0.055037</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19641 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CD-coefficient  Significance\n",
       "gene                                    \n",
       "H19               0.540918      0.064525\n",
       "Col6a3            0.355815      0.055393\n",
       "Acan              0.233381      0.007798\n",
       "Igf2              0.148825      0.061754\n",
       "Col2a1            0.134666      0.055251\n",
       "...                    ...           ...\n",
       "Serpinh1         -0.111309      0.011136\n",
       "RGD1566401       -0.116253      0.012981\n",
       "Col9a1           -0.131756      0.055465\n",
       "Col27a1          -0.136371      0.055230\n",
       "Col11a1          -0.441890      0.055037\n",
       "\n",
       "[19641 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute signatures\n",
    "signature = cd_signature(control_name, treatment, dataset, 'rawdata+filter_genes', meta_class_column_name)\n",
    "signature['rawdata+filter_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14257400",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature['rawdata+filter_genes'].to_csv(\"GSE186104.txt\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
