<!DOCTYPE html>
<html lang="en">
    <head>
        <!-- http://getbootstrap.com/docs/5.1/ -->
        <link crossorigin="anonymous" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" rel="stylesheet">
        <script crossorigin="anonymous" src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p"></script>

        <!--Styling-->
        <link href="styles.css" rel="stylesheet">
        <link href="favicon.ico" rel="icon">

        <title>Dexamethasone Benchmark</title>
    </head>
    <body>
        <nav class="border navbar navbar-expand-md navbar-light">
            <div class="container-fluid" id="nav-text">
                <img alt="logo" src="logo.png" style="height: 3.5rem;">
                <a class="navbar-brand" href="../index.html" style="color: #ffffff">Dexamethasone Benchmarking Resource</a>
                <button aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler" data-bs-target="#navbar" data-bs-toggle="collapse" type="button">
                    <span class="navbar-toggler-icon"></span>
                </button>
                <div class="collapse navbar-collapse" id="navbar">
                    <ul class="navbar-nav ms-auto">
                        <li class="nav-item"><a class="nav-link" href="about.html" style="color: #ffffff">About</a></li>
                        <li class="nav-item"><a class="nav-link" href="code.html" style="color: #ffffff">Code</a></li>
                        <li class="nav-item"><a class="nav-link" href="data.html" style="color: #ffffff">Data</a></li>
                        <li class="nav-item"><a class="nav-link" href="examples.html" style="color: #ffffff">Examples</a></li>
                        <li class="nav-item"><a class="nav-link" href="stats.html" style="color: #ffffff">Stats</a></li>
                    </ul>
                </div>
            </div>
        </nav>
        <main class="py-4">
            <div class="box">
                <h2>
                    Benchmarking Methods for Analysis of RNA-seq Data
                </h2>
                <p>
                    The code notebooks provided on this resource allow users to apply multiple normalization, differential gene expression, and enrichment analysis methods to an RNA-seq dataset. Here, we provide example results and visualizations from the analysis of the study <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE137535" target="_blank">GSE137535</a>. 
                </p>
                <ul>
                    <li>3 control samples (MDA-MB-231 cell line with no drug treatment)</li>
                    <li>3 treatment samples (MDA-MB-231 cell line treated with 1 uM dexamethasone for 24 hours)</li>
                </ul>
                <h3>
                    Differential Gene Expression Methods for RNA-seq Data
                </h3>
                <p>
                    The Brownian bridge plots below show the average rankings of genes from NR3C1, NR0B1, and NR1I2 target gene sets in dexamethasone perturbation gene expression signatures computed using each method. The x-axis represents the gene rank within the signature normalized to [0,1], while the y-axis provides the difference from the uniform cumulative distribution function (CDF), <i>D(r)-r</i>. For more information on Brownian bridge plot visualizations, please refer to our paper on the characteristic direction method (<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-79" target="_blank">Clark et al. 2014</a>). 
                </p>
                <p>
                    We would expect the target genes of transcription factors bound by dexamethasone to show significant differential expression under dexamethasone perturbation. Because target genes may have increased or decreased expression, gene signatures are ranked by the following values for each method:
                </p>
                <ul>
                    <li><b>Characteristic Direction:</b> Absolute value of the CD-coefficient</li>
                    <li><b>Limma-voom:</b> Adjusted p-value</li>
                    <li><b>pyDESeq2:</b> Adjusted p-value</li>
                    <li><b>Welch's t-test:</b> p-value</li>
                    <li><b>Log fold change:</b> Absolute value of the log fold change</li>
                    <li><b>Random:</b> Randomly shuffled gene order</li>
                </ul>
                <div class="row">
                    <div>
                        <img alt="Brownian bridge plot showing average rank of NR3C1 terms for each method" src="plots/rnaseq_deg/rna_dge_nr3c1_bridgeplot.png" width="90%">  
                    </div>
                    <div>
                        <img alt="Leading edge of Brownian bridge plot showing average rank of NR3C1 terms for each method" src="plots/rnaseq_deg/rna_dge_nr3c1_bridgeplot_leadingedge.png" width="60%">  
                    </div>
                </div>
                <p>
                    Above, <b>limma-voom</b> (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29" target="_blank">Law et al. 2014</a>; <a href="https://academic.oup.com/nar/article/43/7/e47/2414268?login=false" target="_blank">Ritchie et al. 2015</a>) and <b>pyDESeq2</b>> (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8", target="_blank">Love et al. 2014</a>; <a href="https://www.biorxiv.org/content/10.1101/2022.12.14.520412v1" target="_blank">Muzellec et al. 2022</a>) recover the most NR3C1 target genes, as shown by their peaks. Additionally, from the leading edge, we can see that these two methods also rank target genes much higher. 
                </p>
                <div class="row">
                    <div>
                        <img alt="Brownian bridge plot showing average rank of NR0B1 terms for each method" src="plots/rnaseq_deg/rna_dge_nr0b1_bridgeplot.png" width="90%">  
                    </div>
                    <div>
                        <img alt="Leading edge of Brownian bridge plot showing average rank of NR0B1 terms for each method" src="plots/rnaseq_deg/rna_dge_nr0b1_bridgeplot_leadingedge.png" width="60%">  
                    </div>
                </div>
                <div class="row">
                    <div>
                        <img alt="Brownian bridge plot showing average rank of NR1I2 terms for each method" src="plots/rnaseq_deg/rna_dge_nr1i2_bridgeplot.png" width="90%">  
                    </div>
                    <div>
                        <img alt="Leading edge of Brownian bridge plot showing average rank of NR1I2 terms for each method" src="plots/rnaseq_deg/rna_dge_nr1i2_bridgeplot_leadingedge.png" width="60%">  
                    </div>
                </div>
                <p>
                    Although NR0B1 and NR1I2 have a weaker drug-target relationship with dexamethasone, a similar pattern emerges in capturing their respective target genes from dexamethasone perturbational signatures. The <b>limma-voom</b> and <b>pyDESeq2</b> signatures consistently display the largest supremum (greatest difference from the uniform CDF), and show a steep leading edge, indicating the recovery of many significantly perturbed genes. The characteristic direction method performs well in the cases of NR0B1 and NR3C1, but less so for NR1I2.
                </p>
            </div>
            <div class="box">
                <h2>
                    References
                </h2>
                <p>Clark, N. R., Hu, K. S., Feldmann, A. S., Kou, Y., Chen, E. Y., Duan, Q., & Ma'ayan, A. (2014). The characteristic direction: a geometrical approach to identify differentially expressed genes. BMC bioinformatics, 15, 79. https://doi.org/10.1186/1471-2105-15-79</p>
                <p>Law, C. W., Chen, Y., Shi, W., & Smyth, G. K. (2014). voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome biology, 15(2), R29. https://doi.org/10.1186/gb-2014-15-2-r29</p>
                <p>Ritchie, M. E., Phipson, B., Wu, D., Hu, Y., Law, C. W., Shi, W., & Smyth, G. K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic acids research, 43(7), e47. https://doi.org/10.1093/nar/gkv007</p>
                <p>Love, M. I., Huber, W., & Anders, S. (2014). Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome biology, 15(12), 550. https://doi.org/10.1186/s13059-014-0550-8</p>
                <p>Muzellec, B., Telenczuk, M., Cabeli, V., & Andreux, M. (2022). PyDESeq2: a python package for bulk RNA-seq differential expression analysis. bioRxiv doi:10.1101/2022.12.14.520412. </p>
            </div>
        </main>
        <footer class="footer">
            <div class="container">
                <div class="row justify-content-center">
                    <div class="col-md-2 col-sm text-center">
                        <a href="https://icahn.mssm.edu/research/bioinformatics" target="_blank">
                            <img class="rounded" src="icahn_cb.png" style="height: 6rem;">
                        </a>
                    </div>
                    <div class="col-md-2 col-sm text-center">
                        <a href="https://labs.icahn.mssm.edu/maayanlab/" target="_blank">
                            <img class="rounded" src="maayanlab_logo.png" style="height: 6rem;">
                        </a>
                    </div>
                    <div class="col-md-2 col-sm">
                        <div class="row my-2" style="padding-top:1rem">
                            <a class="badge badge-secondary px-2" href="https://github.com/MaayanLab/dex-benchmark" target="_blank">
                                View source code
                            </a>
                        </div>
                        <div class="row my-2">
                            <a class="badge badge-secondary px-2"
                               href="https://github.com/MaayanLab/dex-benchmark/issues/new" target="_blank">
                                Submit an issue
                            </a>
                        </div>
                    </div>
                </div>
            </div>
        </footer>
    </body>
</html>